<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Absolute Copy-Number Diversity Heatmap — ACNDiversityPlot • utanos</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Absolute Copy-Number Diversity Heatmap — ACNDiversityPlot"><meta name="description" content="A whole genome heatmap of copy-number changes. It expects absolute copy-numbers.
Default behaviour is to perform hierarchical clustering of the samples and plot the heatmap in that order.
Setting the order parameter overrules this behaviour."><meta property="og:description" content="A whole genome heatmap of copy-number changes. It expects absolute copy-numbers.
Default behaviour is to perform hierarchical clustering of the samples and plot the heatmap in that order.
Setting the order parameter overrules this behaviour."><meta property="og:image" content="https://huntsmanlab.github.io/utanos/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">utanos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/utanos.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/A-samplequality-vignette.html">1. Sample Filtering and Quality Evaluation</a></li>
    <li><a class="dropdown-item" href="../articles/B-cndiversityandscaling-vignette.html">2. Investigating Copy-Number Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/C-cnsignatures-vignette.html">3. Investigating and Creating Copy-Number Signatures</a></li>
    <li><a class="dropdown-item" href="../articles/D-shallowhrd-vignette.html">4. Predicting Homologous Recombination Deficiency (HRD)</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Absolute Copy-Number Diversity Heatmap</h1>

      <div class="d-none name"><code>ACNDiversityPlot.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A whole genome heatmap of copy-number changes. It expects absolute copy-numbers.
Default behaviour is to perform hierarchical clustering of the samples and plot the heatmap in that order.
Setting the <code>order</code> parameter overrules this behaviour.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ACNDiversityPlot</span><span class="op">(</span></span>
<span>  long_segments <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  order <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ret_order <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  save_path <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  rmblacklist <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  addtitle <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sample_labels <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ann_df <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  obj_name <span class="op">=</span> <span class="st">"version1"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-long-segments">long_segments<a class="anchor" aria-label="anchor" href="#arg-long-segments"></a></dt>
<dd><p>A dataframe. This should be in long format with the same number of rows (bins) for each sample. <br>
If using the output from the <code>CalculateACNs</code> function then first run <code>SegmentsToCopyNumber</code> to convert to long. <br>
Expects the following columns: <br>
c("chromosome", "start", "end", "state", "sample_id")</p></dd>


<dt id="arg-order">order<a class="anchor" aria-label="anchor" href="#arg-order"></a></dt>
<dd><p>(optional) Character vector. Defines the order in which samples will be plotted. <br>
This is particularly useful when plotting more than 1 heatmap next to one another. <br>
Allows samples to line-up horizontally.
Example: <br><code>c("CC-CHM-1341", "CC-CHM-1347", "CC-CHM-1355", CC-CHM-1361", "CC-HAM-0369", "CC-HAM-0374", "CC-HAM-0379", "CC-HAM-0383", "CC-HAM-0385")</code></p></dd>


<dt id="arg-ret-order">ret_order<a class="anchor" aria-label="anchor" href="#arg-ret-order"></a></dt>
<dd><p>(optional) Logical. If set to TRUE the function returns the order in which samples were plotted.</p></dd>


<dt id="arg-save-path">save_path<a class="anchor" aria-label="anchor" href="#arg-save-path"></a></dt>
<dd><p>(optional) String. Expects a path to a directory where the plot should be saved. (png)</p></dd>


<dt id="arg-rmblacklist">rmblacklist<a class="anchor" aria-label="anchor" href="#arg-rmblacklist"></a></dt>
<dd><p>optional <em>NULL or character vector</em>. Filter the plot of blacklist regions from this genome. ex. 'hg19'</p></dd>


<dt id="arg-addtitle">addtitle<a class="anchor" aria-label="anchor" href="#arg-addtitle"></a></dt>
<dd><p>(optional) Logical. If set to TRUE add a basic title to the plot.</p></dd>


<dt id="arg-sample-labels">sample_labels<a class="anchor" aria-label="anchor" href="#arg-sample-labels"></a></dt>
<dd><p>(optional) Logical. If set to TRUE, display sample names and y-axis ticks.</p></dd>


<dt id="arg-ann-df">ann_df<a class="anchor" aria-label="anchor" href="#arg-ann-df"></a></dt>
<dd><p>(optional) Dataframe. Expects the same sample ids as passed to long_segments param. <br>
See the roxygen docs for <code><a href="PlotAnnotationBars.html">PlotAnnotationBars()</a></code> for more details.</p></dd>


<dt id="arg-obj-name">obj_name<a class="anchor" aria-label="anchor" href="#arg-obj-name"></a></dt>
<dd><p>(optional) String. Adds a tag to the end of the filename if saving the image. <br>
Only used if the save_path parameter is also set.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A ggplot2 object or a list of a ggplot2 object and vector of the ordered sample names.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Maxwell Douglas, Branden Lynch, Carlos Vasquez, Jacky Yiu, Matthew Eldridge, Geoff Macintyre, Ding Ma, Nirupama Tamvada, Sameer Shankar.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

