<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Calculate Absolute Copy Numbers — CalculateACNs • utanos</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate Absolute Copy Numbers — CalculateACNs"><meta name="description" content="CalculateACNs() calculates absolute copy numbers (ACNs) from the relative copy number profiles of one or more samples.
There are several included options by which to do this.
Note: If not providing a table of variant allele frequencies (VAFs) then 'mad' is the only method available.
This function makes use of the rascal package in R and instructions can be found in the vignette:
https://github.com/crukci-bioinformatics/rascal/blob/master/vignettes/rascal.Rmd  "><meta property="og:description" content="CalculateACNs() calculates absolute copy numbers (ACNs) from the relative copy number profiles of one or more samples.
There are several included options by which to do this.
Note: If not providing a table of variant allele frequencies (VAFs) then 'mad' is the only method available.
This function makes use of the rascal package in R and instructions can be found in the vignette:
https://github.com/crukci-bioinformatics/rascal/blob/master/vignettes/rascal.Rmd  "><meta property="og:image" content="https://huntsmanlab.github.io/utanos/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">utanos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/utanos.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/A-samplequality-vignette.html">1. Sample Filtering and Quality Evaluation</a></li>
    <li><a class="dropdown-item" href="../articles/B-cndiversityandscaling-vignette.html">2. Investigating Copy-Number Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/C-cnsignatures-vignette.html">3. Investigating and Creating Copy-Number Signatures</a></li>
    <li><a class="dropdown-item" href="../articles/D-shallowhrd-vignette.html">4. Predicting Homologous Recombination Deficiency (HRD)</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Calculate Absolute Copy Numbers</h1>

      <div class="d-none name"><code>CalculateACNs.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>CalculateACNs() calculates absolute copy numbers (ACNs) from the relative copy number profiles of one or more samples.
There are several included options by which to do this.
Note: If not providing a table of variant allele frequencies (VAFs) then 'mad' is the only method available. <br></p>
<p>This function makes use of the rascal package in R and instructions can be found in the vignette: <br>
https://github.com/crukci-bioinformatics/rascal/blob/master/vignettes/rascal.Rmd  <br></p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">CalculateACNs</span><span class="op">(</span></span>
<span>  <span class="va">cnobj</span>,</span>
<span>  <span class="va">acnmethod</span>,</span>
<span>  rascal_sols <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  variants <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  acn_save_path <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  return_sols <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  return_S4 <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  distance_decimal_places <span class="op">=</span> <span class="fl">7</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-cnobj">cnobj<a class="anchor" aria-label="anchor" href="#arg-cnobj"></a></dt>
<dd><p>An S4 object of type QDNAseqCopyNumbers. This object must contain a copynumber slot and a segmented slot.</p></dd>


<dt id="arg-acnmethod">acnmethod<a class="anchor" aria-label="anchor" href="#arg-acnmethod"></a></dt>
<dd><p>The method by which to calculate ACNs. Can be one of: <br>
"maxvaf" - Calculate ACNs assuming the maximum discovered VAF for the sample is an appropriate representation for the tumour fraction. <br>
char vector - Same as above but rather than using the max vaf provide a character vector of the genes from which to pull VAFs.
The genes are assumed to be in order of decreasing precedence. ex. c('TP53', 'KRAS', 'PTEN') <br>
"mad" - Calculate ACNs using the mean absolute difference (MAD) column from the solutions table. <br>
If using variant allele frequencies from targeted panel sequencing or some other technology: <br></p><ul><li><p>The variants must must be in a datatable/dataframe. <br></p></li>
<li><p>Required columns: sample_id, chromosome, start, end, gene_name, ref, alt, vaf.</p></li>
<li><p>Each row of said table must correspond to a unique variant. <br></p></li>
<li><p>Each variant must have an associated variant allele frequency. <br></p></li>
<li><p>Each row must also be associated with a specific sample. <br>
A dataframe of known ploidies per sample - Calculate ACNs at each bin given we already know the ploidy of the samples. <br></p></li>
</ul></dd>


<dt id="arg-rascal-sols">rascal_sols<a class="anchor" aria-label="anchor" href="#arg-rascal-sols"></a></dt>
<dd><p>A tsv or dataframe of the calculated rascal solutions.</p></dd>


<dt id="arg-variants">variants<a class="anchor" aria-label="anchor" href="#arg-variants"></a></dt>
<dd><p>(optional) Dataframe or string. A table of variants containing variant allele frequencies per gene and per sample.</p></dd>


<dt id="arg-acn-save-path">acn_save_path<a class="anchor" aria-label="anchor" href="#arg-acn-save-path"></a></dt>
<dd><p>(optional) String. The output path (absolute path recommended) where to save the result.</p></dd>


<dt id="arg-return-sols">return_sols<a class="anchor" aria-label="anchor" href="#arg-return-sols"></a></dt>
<dd><p>(optional) Logical. Return the selected rascal solution.</p></dd>


<dt id="arg-return-s-">return_S4<a class="anchor" aria-label="anchor" href="#arg-return-s-"></a></dt>
<dd><p>(optional) Logical.</p></dd>


<dt id="arg-distance-decimal-places">distance_decimal_places<a class="anchor" aria-label="anchor" href="#arg-distance-decimal-places"></a></dt>
<dd><p>number of decimal places to compare to when finding
the minimum distance.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing: <br></p><ol><li><p>A list of dataframes (one for each sample) OR a QDNAseq S4 object. <br></p></li>
<li><p>Optionally, a dataframe of the rascal solutions. <br>
This function returns ACNs when a rascal solution is found, if one isn't, the sample is skipped.</p></li>
</ol></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p></p><div class="sourceCode"><pre><code><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>solutions <span class="ot">&lt;-</span> <span class="st">"~/Documents/.../rascal_solutions.csv"</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>rcn.obj <span class="ot">&lt;-</span> <span class="st">"~/repos/utanosmodellingdata/sample_copy_number_data/sample_filtered_cn_data.rds"</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>variants <span class="ot">&lt;-</span> <span class="st">"~/Documents/.../allvariants.clinvar.cosmic.exons.csv"</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>save_path <span class="ot">&lt;-</span> <span class="st">"~/Documents/.../rascal_ACN_segments.rds"</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>variants <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="at">file =</span> variants, <span class="at">sep =</span> <span class="st">','</span>)</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>variants <span class="ot">&lt;-</span> variants <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">chromosome =</span> chr,</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>                                       <span class="at">gene_name =</span> genecode_gene_name)</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>outputs <span class="ot">&lt;-</span> <span class="fu">CalculateACNs</span>(rcn.obj,</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>                         <span class="at">acn.method =</span> <span class="st">'maxvaf'</span>,</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>                         <span class="at">rascal_sols =</span> solutions,</span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>                         <span class="at">variants =</span> variants,</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a>                         <span class="at">return_sols =</span> <span class="cn">TRUE</span>,</span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a>                         <span class="at">return_S4 =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">CalculateACNs</span>(rcn.obj,</span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a>                        <span class="at">rascal_sols =</span> solutions,</span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a>                        <span class="at">variants =</span> variants,</span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a>                        <span class="at">acnmethod =</span> <span class="fu">c</span>(<span class="st">'TP53'</span>, <span class="st">'KRAS'</span>, <span class="st">'BRCA1'</span>,</span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a>                                      <span class="st">'BRCA2'</span>, <span class="st">'PIK3CA'</span>, <span class="st">'PTEN'</span>),</span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a>                        <span class="at">acn_save_path =</span> save_path)</span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">CalculateACNs</span>(rcn.obj,</span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a>                        <span class="at">rascal_sols =</span> solutions,</span>
<span id="cb1-22"><a href="#cb1-22" tabindex="-1"></a>                        <span class="at">acnmethod =</span> <span class="st">'mad'</span>)</span>
<span id="cb1-23"><a href="#cb1-23" tabindex="-1"></a></span></code></pre><p></p></div>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Maxwell Douglas, Branden Lynch, Carlos Vasquez, Jacky Yiu, Matthew Eldridge, Geoff Macintyre, Ding Ma, Nirupama Tamvada, Sameer Shankar.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

