<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Segment Copy Number Plot across the Genome — CNSegmentsPlot • utanos</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Segment Copy Number Plot across the Genome — CNSegmentsPlot"><meta name="description" content="Whole genome copy number plot for one sample with partitioning lines between chromosomes."><meta property="og:description" content="Whole genome copy number plot for one sample with partitioning lines between chromosomes."><meta property="og:image" content="https://huntsmanlab.github.io/utanos/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">utanos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/utanos.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/A-samplequality-vignette.html">1. Sample Filtering and Quality Evaluation</a></li>
    <li><a class="dropdown-item" href="../articles/B-cndiversityandscaling-vignette.html">2. Investigating Copy-Number Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/C-cnsignatures-vignette.html">3. Investigating and Creating Copy-Number Signatures</a></li>
    <li><a class="dropdown-item" href="../articles/D-shallowhrd-vignette.html">4. Predicting Homologous Recombination Deficiency (HRD)</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Segment Copy Number Plot across the Genome</h1>

      <div class="d-none name"><code>CNSegmentsPlot.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Whole genome copy number plot for one sample with partitioning lines between chromosomes.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">CNSegmentsPlot</span><span class="op">(</span></span>
<span>  <span class="va">cnobj</span>,</span>
<span>  sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  copy_number_steps <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  max_points_to_display <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  highlight_masks <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_copy_number <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  max_copy_number <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  copy_number_breaks <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dolog2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  def_point_colour <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  point_alpha <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>  point_size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  point_shape <span class="op">=</span> <span class="fl">16</span>,</span>
<span>  segment_colour <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>  segment_alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  segment_line_size <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>  copy_number_step_colour <span class="op">=</span> <span class="st">"blue"</span>,</span>
<span>  copy_number_step_alpha <span class="op">=</span> <span class="fl">0.35</span>,</span>
<span>  copy_number_step_line_size <span class="op">=</span> <span class="fl">0.75</span>,</span>
<span>  xlabel <span class="op">=</span> <span class="st">"chromosome"</span>,</span>
<span>  ylabel <span class="op">=</span> <span class="st">"copy number"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-cnobj">cnobj<a class="anchor" aria-label="anchor" href="#arg-cnobj"></a></dt>
<dd><p>A <em>QDNAseq</em> object contains <code>copynumber</code>, and <code>segmented</code> assay slots. <br>
This function expects copy-number data that has been segmented. May contain multiple samples.</p></dd>


<dt id="arg-sample">sample<a class="anchor" aria-label="anchor" href="#arg-sample"></a></dt>
<dd><p>A <em>character string</em> indicating the sample. <br>
Required if the object passed to <code>cnobj</code> contains data for multiple samples. <br>
Ex. <code>c('sample1')</code> <br>
Must match one of the sample names exactly.</p></dd>


<dt id="arg-copy-number-steps">copy_number_steps<a class="anchor" aria-label="anchor" href="#arg-copy-number-steps"></a></dt>
<dd><p>a data frame containing <code>absolute_copy_number</code>
and <code>copy_number</code> columns.</p></dd>


<dt id="arg-max-points-to-display">max_points_to_display<a class="anchor" aria-label="anchor" href="#arg-max-points-to-display"></a></dt>
<dd><p>maximum number of copy number points to display
(downsampling may be carried out if there are more copy number values than
this number).</p></dd>


<dt id="arg-highlight-masks">highlight_masks<a class="anchor" aria-label="anchor" href="#arg-highlight-masks"></a></dt>
<dd><p>A <em>character string</em> indicating any feature data slots
containing logical masks over the bins. Each slot will be given its own colour. <br>
Keep in mind that NA values will be filtered out and should masks have overlap
the last one will take precedence. <br>
Ex. <code>c('comCNV.mask', 'centro.telo.mask')</code></p></dd>


<dt id="arg-min-copy-number">min_copy_number<a class="anchor" aria-label="anchor" href="#arg-min-copy-number"></a></dt>
<dd><p>Numeric. The minimum <code>copy_number</code> to display. Effectively controls Y-axis limits.</p></dd>


<dt id="arg-max-copy-number">max_copy_number<a class="anchor" aria-label="anchor" href="#arg-max-copy-number"></a></dt>
<dd><p>Numeric. The maximum <code>copy_number</code> to display. Effectively controls Y-axis limits.</p></dd>


<dt id="arg-copy-number-breaks">copy_number_breaks<a class="anchor" aria-label="anchor" href="#arg-copy-number-breaks"></a></dt>
<dd><p>Vector of integers. Breaks at which grid lines will be displayed. <br>
ex. c(1:12)</p></dd>


<dt id="arg-def-point-colour">def_point_colour<a class="anchor" aria-label="anchor" href="#arg-def-point-colour"></a></dt>
<dd><p>A <em>character string</em> indicating the colour for the
default copy number points.</p></dd>


<dt id="arg-point-alpha">point_alpha<a class="anchor" aria-label="anchor" href="#arg-point-alpha"></a></dt>
<dd><p>Numeric. The transparency of the copy number points.</p></dd>


<dt id="arg-point-size">point_size<a class="anchor" aria-label="anchor" href="#arg-point-size"></a></dt>
<dd><p>The size of the copy number points.</p></dd>


<dt id="arg-point-shape">point_shape<a class="anchor" aria-label="anchor" href="#arg-point-shape"></a></dt>
<dd><p>The point shape. See the ggplot2::geom_point() docs and
the <code>shape</code> parameter for details.</p></dd>


<dt id="arg-segment-colour">segment_colour<a class="anchor" aria-label="anchor" href="#arg-segment-colour"></a></dt>
<dd><p>The colour of the copy number segments.</p></dd>


<dt id="arg-segment-alpha">segment_alpha<a class="anchor" aria-label="anchor" href="#arg-segment-alpha"></a></dt>
<dd><p>The transparency of the copy number segments.</p></dd>


<dt id="arg-segment-line-size">segment_line_size<a class="anchor" aria-label="anchor" href="#arg-segment-line-size"></a></dt>
<dd><p>The size of the lines for copy number segments.</p></dd>


<dt id="arg-copy-number-step-colour">copy_number_step_colour<a class="anchor" aria-label="anchor" href="#arg-copy-number-step-colour"></a></dt>
<dd><p>The colour of the copy number step lines.</p></dd>


<dt id="arg-copy-number-step-alpha">copy_number_step_alpha<a class="anchor" aria-label="anchor" href="#arg-copy-number-step-alpha"></a></dt>
<dd><p>The transparency of the copy number step lines.</p></dd>


<dt id="arg-copy-number-step-line-size">copy_number_step_line_size<a class="anchor" aria-label="anchor" href="#arg-copy-number-step-line-size"></a></dt>
<dd><p>The size of the lines for the copy number.
steps.</p></dd>


<dt id="arg-xlabel-ylabel">xlabel, ylabel<a class="anchor" aria-label="anchor" href="#arg-xlabel-ylabel"></a></dt>
<dd><p>x- and y-axis labels.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>ggplot</code> object.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The <code>copynumber</code> values passed into this function can be on the relative
or absolute scales, or be log2 ratios. However, the same scale should be used
consistently in each slot of <code>cnobj</code> (and the <code>copy_number_steps</code> dataframe as the case may be).</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Maxwell Douglas, Branden Lynch, Carlos Vasquez, Jacky Yiu, Matthew Eldridge, Geoff Macintyre, Ding Ma, Nirupama Tamvada, Sameer Shankar.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

