<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Extract Relative Copy-Number Features — ExtractRelativeCopyNumberFeatures • utanos</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Extract Relative Copy-Number Features — ExtractRelativeCopyNumberFeatures"><meta name="description" content="Extract genome-wide copy-number features from either a list of dataframes or a QDNAseq S4 object for 1 or more samples.
This function is intended to be run on relative CN data."><meta property="og:description" content="Extract genome-wide copy-number features from either a list of dataframes or a QDNAseq S4 object for 1 or more samples.
This function is intended to be run on relative CN data."><meta property="og:image" content="https://huntsmanlab.github.io/utanos/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">utanos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/utanos.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/0_quickstart-vignette.html">0. utanos quickstart guide</a></li>
    <li><a class="dropdown-item" href="../articles/1_samplequality-vignette.html">1. Sample Filtering and Quality Evaluation</a></li>
    <li><a class="dropdown-item" href="../articles/2_cndiversityandscaling-vignette.html">2. Investigating Copy-Number Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/3_cnsignatures-vignette.html">3. Investigating and Creating Copy-Number Signatures</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Extract Relative Copy-Number Features</h1>

      <div class="d-none name"><code>ExtractRelativeCopyNumberFeatures.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Extract genome-wide copy-number features from either a list of dataframes or a QDNAseq S4 object for 1 or more samples.
This function is intended to be run on relative CN data.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ExtractRelativeCopyNumberFeatures</span><span class="op">(</span></span>
<span>  <span class="va">CN_data</span>,</span>
<span>  <span class="va">genome</span>,</span>
<span>  cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  log_gaussians <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  extra_features <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-cn-data">CN_data<a class="anchor" aria-label="anchor" href="#arg-cn-data"></a></dt>
<dd><p>List of datafames or S4 QDNAseq object. Segmented relative copy-number data for 1 or more samples.
If input is a list of dataframes, columns should be: <br></p><ol><li><p>chromosome</p></li>
<li><p>start</p></li>
<li><p>end</p></li>
<li><p>segVal</p></li>
</ol></dd>


<dt id="arg-genome">genome<a class="anchor" aria-label="anchor" href="#arg-genome"></a></dt>
<dd><p>Character string. The reference genome used for alignment. <br>
Options: 'hg19', 'hg38'</p></dd>


<dt id="arg-cores">cores<a class="anchor" aria-label="anchor" href="#arg-cores"></a></dt>
<dd><p>Integer. The number of cores to use for parallel processing.</p></dd>


<dt id="arg-log-gaussians">log_gaussians<a class="anchor" aria-label="anchor" href="#arg-log-gaussians"></a></dt>
<dd><p>Logical. If TRUE, take the log1p of the extracted CN-features for those being modeled by gaussians. (segsize, changepoint, copynumber)</p></dd>


<dt id="arg-extra-features">extra_features<a class="anchor" aria-label="anchor" href="#arg-extra-features"></a></dt>
<dd><p>Logical. If TRUE, extracts CN-feature data for two more features: nc50, and cdist.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list. Each list element contains feature data for a single feature.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function is identical to the absolute calling equivalent other than for three features.
The <code>osCN</code>, <code>changepoint</code>, and <code>copynumber</code> features require slightly different modelling at the relative scale.</p>
<p>The extracted copy-number features are:</p><ol><li><p>Breakpoint count per 10MB - <code>bp10MB</code></p></li>
<li><p>Copy-number value of each segment - <code>copynumber</code></p></li>
<li><p>Copy-number difference between adjacent segments - <code>changepoint</code></p></li>
<li><p>Breakpoint count per chromosome arm - <code>bpchrarm</code></p></li>
<li><p>Lengths of oscillating CN segment chains - <code>osCN</code></p></li>
<li><p>Size of copy-number segments in base-pairs - <code>segsize</code></p></li>
</ol><p>Extra features: <br>
7. Minimum number of chromosomes (a count) needed to account for 50% of CN changes in a sample - <code>nc50</code> <br>
8. Distance in base pairs of each breakpoint to the centromere - <code>cdist</code></p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Maxwell Douglas, Branden Lynch, Carlos Vasquez, Jacky Yiu, Ding Ma, Nirupama Tamvada, Sameer Shankar.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

