<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ utanos</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://huntsmanlab.github.io/utanos/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">utanos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/utanos.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/A-samplequality-vignette.html">1. Sample Filtering and Quality Evaluation</a></li>
    <li><a class="dropdown-item" href="../articles/B-cndiversityandscaling-vignette.html">2. Investigating Copy-Number Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/C-cnsignatures-vignette.html">3. Investigating and Creating Copy-Number Signatures</a></li>
    <li><a class="dropdown-item" href="../articles/D-shallowhrd-vignette.html">4. Predicting Homologous Recombination Deficiency (HRD)</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="ACNDiversityPlot.html">ACNDiversityPlot()</a></code>

        </dt>
        <dd>Absolute Copy-Number Diversity Heatmap</dd>
      </dl><dl><dt>

          <code><a href="AbsoluteCopyNumberDistance.html">AbsoluteCopyNumberDistance()</a></code>

        </dt>
        <dd>Compute distance function for fitting relative copy numbers to absolute</dd>
      </dl><dl><dt>

          <code><a href="AbsoluteCopyNumberDistanceGrid.html">AbsoluteCopyNumberDistanceGrid()</a></code>

        </dt>
        <dd>Compute distance function for fitting relative copy numbers for a grid of ploidies and cellularities</dd>
      </dl><dl><dt>

          <code><a href="AddChromosomeArm.html">AddChromosomeArm()</a></code>

        </dt>
        <dd>Adds a column for the chromosome arm in which each segment is in.</dd>
      </dl><dl><dt>

          <code><a href="AddChromosomeArmHelper.html">AddChromosomeArmHelper()</a></code>

        </dt>
        <dd>Adds a chromosome arm column to df at the respective positions for each chromosome.</dd>
      </dl><dl><dt>

          <code><a href="AddGenesToPlot.html">AddGenesToPlot()</a></code>

        </dt>
        <dd>Label genes of interest on an absolute copy-number plot.</dd>
      </dl><dl><dt>

          <code><a href="AssignLevels.html">AssignLevels()</a></code>

        </dt>
        <dd>Gather big segments and assign levels iteratively</dd>
      </dl><dl><dt>

          <code><a href="BreakSmoothToLGA.html">BreakSmoothToLGA()</a></code>

        </dt>
        <dd>Preps <code>segments</code> for LGA call.</dd>
      </dl><dl><dt>

          <code><a href="BuildWxQdnaObject.html">BuildWxQdnaObject()</a></code>

        </dt>
        <dd>Build a QDNAseq object from the WisecondorX output files</dd>
      </dl><dl><dt>

          <code><a href="CNSegmentsPlot.html">CNSegmentsPlot()</a></code>

        </dt>
        <dd>Segment Copy Number Plot across the Genome</dd>
      </dl><dl><dt>

          <code><a href="CalculateACNs.html">CalculateACNs()</a></code>

        </dt>
        <dd>Calculate Absolute Copy Numbers</dd>
      </dl><dl><dt>

          <code><a href="CallLGA.html">CallLGA()</a></code>

        </dt>
        <dd>Determines number of large genomic alterations (LGAs) for sizes 3:11</dd>
      </dl><dl><dt>

          <code><a href="CallSignatureExposures.html">CallSignatureExposures()</a></code>

        </dt>
        <dd>Wrapper to find Signature Exposures for a set of samples</dd>
      </dl><dl><dt>

          <code><a href="ChooseNumberSignatures.html">ChooseNumberSignatures()</a></code>

        </dt>
        <dd>Choose an Optimal Number of Signatures</dd>
      </dl><dl><dt>

          <code><a href="ChromosomeSplitPos.html">ChromosomeSplitPos()</a></code>

        </dt>
        <dd>Split Chromosome Name</dd>
      </dl><dl><dt>

          <code><a href="CleanBamRatiosFrame.html">CleanBamRatiosFrame()</a></code>

        </dt>
        <dd>Performs some cleaning operations.</dd>
      </dl><dl><dt>

          <code><a href="CollapseTableToSegments.html">CollapseTableToSegments()</a></code>

        </dt>
        <dd>Collapses down table to segmented copy numbers</dd>
      </dl><dl><dt>

          <code><a href="CollapsedSegs.html">CollapsedSegs()</a></code>

        </dt>
        <dd>Collapse segments</dd>
      </dl><dl><dt>

          <code><a href="CopyNumberSegments.html">CopyNumberSegments()</a></code>

        </dt>
        <dd>Collapses relative copy-number calls to segment tables</dd>
      </dl><dl><dt>

          <code><a href="CopySegFlat.html">CopySegFlat()</a></code>

        </dt>
        <dd>Make a DF of CNs and Segments from S4 QDNAseq object</dd>
      </dl><dl><dt>

          <code><a href="CorrectLeftovers.html">CorrectLeftovers()</a></code>

        </dt>
        <dd>Adds 'leftsovers' to the segment frame after re-inserting small segments.</dd>
      </dl><dl><dt>

          <code><a href="DetermineNumberOfLGAs.html">DetermineNumberOfLGAs()</a></code>

        </dt>
        <dd>Determines if the given segments are LGAs of the given size</dd>
      </dl><dl><dt>

          <code><a href="DlpToQDNAseq.html">DlpToQDNAseq()</a></code>

        </dt>
        <dd>Build a QDNAseq object from the DLP output files</dd>
      </dl><dl><dt>

          <code><a href="ExcludeShortArms.html">ExcludeShortArms()</a></code>

        </dt>
        <dd>Removes the short chromosome arms (21, 22, 24) &amp; keeps segments with size &gt;= 3Mb</dd>
      </dl><dl><dt>

          <code><a href="ExportBinsQDNAObj.html">ExportBinsQDNAObj()</a></code>

        </dt>
        <dd>Export the CN bins of a QDNAseq Object to a wide dataframe</dd>
      </dl><dl><dt>

          <code><a href="ExtractCopyNumberFeatures.html">ExtractCopyNumberFeatures()</a></code>

        </dt>
        <dd>Extract Absolute Copy-Number Features</dd>
      </dl><dl><dt>

          <code><a href="ExtractRelativeCopyNumberFeatures.html">ExtractRelativeCopyNumberFeatures()</a></code>

        </dt>
        <dd>Extract Relative Copy-Number Features</dd>
      </dl><dl><dt>

          <code><a href="FilterCNs.html">FilterCNs()</a></code>

        </dt>
        <dd>Remove or set a mask for relative CN bins (QDNAseq object)</dd>
      </dl><dl><dt>

          <code><a href="FinalizeSmallSegments.html">FinalizeSmallSegments()</a></code>

        </dt>
        <dd>Re-inserts small segments in between the large ones.</dd>
      </dl><dl><dt>

          <code><a href="FindBestFitSolutions.html">FindBestFitSolutions()</a></code>

        </dt>
        <dd>Find best fit solutions for fitting copy numbers through grid-based search</dd>
      </dl><dl><dt>

          <code><a href="FindBigSegmentsIndices.html">FindBigSegmentsIndices()</a></code>

        </dt>
        <dd>Finds the indices of big segments that are close ratio_median-wise</dd>
      </dl><dl><dt>

          <code><a href="FindRascalSolutions.html">FindRascalSolutions()</a></code>

        </dt>
        <dd>Find best fitting solutions for each sample</dd>
      </dl><dl><dt>

          <code><a href="FindThreshold.html">FindThreshold()</a></code>

        </dt>
        <dd>Determines threshold for ratio_median difference via KDE.</dd>
      </dl><dl><dt>

          <code><a href="FitComponent.html">FitComponent()</a></code>

        </dt>
        <dd>Fit a Mixture Model for a provided CN-Feature</dd>
      </dl><dl><dt>

          <code><a href="FitMixtureModels.html">FitMixtureModels()</a></code>

        </dt>
        <dd>Fit Mixture Models for each CN-Feature</dd>
      </dl><dl><dt>

          <code><a href="GatherSegmentsByLevels.html">GatherSegmentsByLevels()</a></code>

        </dt>
        <dd>Merges segments by level.</dd>
      </dl><dl><dt>

          <code><a href="GatherSegmentsByRatioMedian.html">GatherSegmentsByRatioMedian()</a></code>

        </dt>
        <dd>Merges segments that are in the same chromosome arm AND have the same ratio_median.</dd>
      </dl><dl><dt>

          <code><a href="GatherSegmentsByRatioMedianHelper.html">GatherSegmentsByRatioMedianHelper()</a></code>

        </dt>
        <dd>Merges segments that have the same ratio_median and are in the same chromosome arm.</dd>
      </dl><dl><dt>

          <code><a href="GaussiansMixturePlot.html">GaussiansMixturePlot()</a></code>

        </dt>
        <dd>Plot a Mixture of Gaussians</dd>
      </dl><dl><dt>

          <code><a href="GenHumanReadableAcnProfile.html">GenHumanReadableAcnProfile()</a></code>

        </dt>
        <dd>Generate Human Readable Absolute Copy-Number Profiles</dd>
      </dl><dl><dt>

          <code><a href="GenHumanReadableRcnProfile.html">GenHumanReadableRcnProfile()</a></code>

        </dt>
        <dd>Generate Human Readable Relative Copy-Number Profiles</dd>
      </dl><dl><dt>

          <code><a href="GenerateSampleByComponentMatrix.html">GenerateSampleByComponentMatrix()</a></code>

        </dt>
        <dd>Generate sum-of-posteriors probability matrix</dd>
      </dl><dl><dt>

          <code><a href="GenerateSignatures.html">GenerateSignatures()</a></code>

        </dt>
        <dd>Create CN-Signatures</dd>
      </dl><dl><dt>

          <code><a href="GetCNTally.html">GetCNTally()</a></code>

        </dt>
        <dd>Extract Copy Number Tally Matrix</dd>
      </dl><dl><dt>

          <code><a href="GetCriticalPoints.html">GetCriticalPoints()</a></code>

        </dt>
        <dd>Finds critical points (minima or maxima) of the differences between segments ratios.</dd>
      </dl><dl><dt>

          <code><a href="GetGRangesObject.html">GetGRangesObject()</a></code>

        </dt>
        <dd>Returns a GRanges object used to obtain data from specified genomic regions.</dd>
      </dl><dl><dt>

          <code><a href="GetHRDStatus.html">GetHRDStatus()</a></code>

        </dt>
        <dd>Returns the HRD status of a sample</dd>
      </dl><dl><dt>

          <code><a href="GetInfoSegm.html">GetInfoSegm()</a></code>

        </dt>
        <dd>TODO: write docs</dd>
      </dl><dl><dt>

          <code><a href="GetLGAOfSize.html">GetLGAOfSize()</a></code>

        </dt>
        <dd>Returns the segments that are LGAs of the given size</dd>
      </dl><dl><dt>

          <code><a href="GetLargeSegments.html">GetLargeSegments()</a></code>

        </dt>
        <dd>Returns segments with a size &gt;= 3Mb.</dd>
      </dl><dl><dt>

          <code><a href="GetSampleQualityDecision.html">GetSampleQualityDecision()</a></code>

        </dt>
        <dd>Calculate and make a quality call for relative copy number profiles</dd>
      </dl><dl><dt>

          <code><a href="GetSegCounts.html">GetSegCounts()</a></code>

        </dt>
        <dd>Calculate segment sizes</dd>
      </dl><dl><dt>

          <code><a href="GetSegmentID.html">GetSegmentID()</a></code>

        </dt>
        <dd>Re-assigns level column values for all segments.</dd>
      </dl><dl><dt>

          <code><a href="GetSegmentationBeforeLGACall.html">GetSegmentationBeforeLGACall()</a></code>

        </dt>
        <dd>Final segmentation before LGA call</dd>
      </dl><dl><dt>

          <code><a href="GetSmallSegments.html">GetSmallSegments()</a></code>

        </dt>
        <dd>Returns the segments with 3Mb &gt;= size &gt;= 0.1Mb</dd>
      </dl><dl><dt>

          <code><a href="InitalizeSmallSegments.html">InitalizeSmallSegments()</a></code>

        </dt>
        <dd>Initializes the process of re-inserting the small segments in between the large ones.</dd>
      </dl><dl><dt>

          <code><a href="InsertSmallSegments.html">InsertSmallSegments()</a></code>

        </dt>
        <dd>Adds small segments in <code>small_segments</code> that meet certain criteria (position wise w.r.t large segments) into the <code>large_segments</code> data frame. Merges if necessary (i.e. if ratio_median difference &lt; threshold).</dd>
      </dl><dl><dt>

          <code><a href="IsAcceptablePloidyAndCellularity.html">IsAcceptablePloidyAndCellularity()</a></code>

        </dt>
        <dd>Determine if a given ploidy and cellularity for fitting copy numbers is acceptable</dd>
      </dl><dl><dt>

          <code><a href="LargeMissingChrArms.html">LargeMissingChrArms()</a></code>

        </dt>
        <dd>Deals with missing chromosome arm values in the given data frame of large segments.</dd>
      </dl><dl><dt>

          <code><a href="MakeSummaryTable.html">MakeSummaryTable()</a></code>

        </dt>
        <dd>Makes Relative Copy-Number Gain/Loss Summary Tables</dd>
      </dl><dl><dt>

          <code><a href="MedSegVar.html">MedSegVar()</a></code>

        </dt>
        <dd>Calculate median segment-level variance per sample</dd>
      </dl><dl><dt>

          <code><a href="MergeSegmentsFiveLargeSmall.html">MergeSegmentsFiveLargeSmall()</a></code>

        </dt>
        <dd>Deals with merging large segment to small segment.</dd>
      </dl><dl><dt>

          <code><a href="MergeSegmentsFiveSmallNextLarge.html">MergeSegmentsFiveSmallNextLarge()</a></code>

        </dt>
        <dd>Deals with merging small segment to next_large segment.</dd>
      </dl><dl><dt>

          <code><a href="MergeSegmentsFour.html">MergeSegmentsFour()</a></code>

        </dt>
        <dd>Handles merging step for the fourth case of FinalizeSmallSegments.</dd>
      </dl><dl><dt>

          <code><a href="MergeSegmentsTwo.html">MergeSegmentsTwo()</a></code>

        </dt>
        <dd>Helper for merging segments in the second case of FinalizeSmallSegments. Returns the updated <code>large_segments</code> data frame with segments appropriately merged (or not... depending on parameters).</dd>
      </dl><dl><dt>

          <code><a href="MixtureModelPlots.html">MixtureModelPlots()</a></code>

        </dt>
        <dd>Plot Mixture Model Components by Signature</dd>
      </dl><dl><dt>

          <code><a href="NullListOfThresholds.html">NullListOfThresholds()</a></code>

        </dt>
        <dd>This function is ran in case that no possible thresholds were found.</dd>
      </dl><dl><dt>

          <code><a href="PlotAnnotationBars.html">PlotAnnotationBars()</a></code>

        </dt>
        <dd>Create annotation bars</dd>
      </dl><dl><dt>

          <code><a href="PlotCNTallySig.html">PlotCNTallySig()</a></code>

        </dt>
        <dd>Plot Copy Number Tally Signature Contributions</dd>
      </dl><dl><dt>

          <code><a href="PlotSegmentChanges.html">PlotSegmentChanges()</a></code>

        </dt>
        <dd>A plot illustrating the merging of segments by the SHRD algorithm.</dd>
      </dl><dl><dt>

          <code><a href="PoissonsMixturePlot.html">PoissonsMixturePlot()</a></code>

        </dt>
        <dd>Plot a Mixture of Poissons</dd>
      </dl><dl><dt>

          <code><a href="PrepFirstRound.html">PrepFirstRound()</a></code>

        </dt>
        <dd>Preps the segment data frame for use in the first round of threshold estimation.</dd>
      </dl><dl><dt>

          <code><a href="PrepForLevelsInitialization.html">PrepForLevelsInitialization()</a></code>

        </dt>
        <dd>Calls some helpers to do something to gathered by ratio medians.... gotta check this Also adds a Level and an Index column to the data frame</dd>
      </dl><dl><dt>

          <code><a href="PrepSecondRound.html">PrepSecondRound()</a></code>

        </dt>
        <dd>Preps the segment data frame for use in the second round of threshold estimation.</dd>
      </dl><dl><dt>

          <code><a href="QualityPlot.html">QualityPlot()</a></code>

        </dt>
        <dd>Plot sample quality according to a specified metric.</dd>
      </dl><dl><dt>

          <code><a href="QuantifySignatures.html">QuantifySignatures()</a></code>

        </dt>
        <dd>Quantify Signature Exposures</dd>
      </dl><dl><dt>

          <code><a href="RCNDiversityPlot.html">RCNDiversityPlot()</a></code>

        </dt>
        <dd>Plot relative CN diversity heatmap</dd>
      </dl><dl><dt>

          <code><a href="RatioDiffBelowThreshold.html">RatioDiffBelowThreshold()</a></code>

        </dt>
        <dd>Determines if absolute ratio_median difference is less than or equal to the given threshold.</dd>
      </dl><dl><dt>

          <code><a href="RelativeToAbsoluteCopyNumber.html">RelativeToAbsoluteCopyNumber()</a></code>

        </dt>
        <dd>Convert relative copy numbers to absolute copy numbers</dd>
      </dl><dl><dt>

          <code><a href="RemoveCentromereTelomeres.html">RemoveCentromereTelomeres()</a></code>

        </dt>
        <dd>Removes centromeres and telomeres from 22 chromosomes (23rd if specified)</dd>
      </dl><dl><dt>

          <code><a href="RemoveSpuriousRegions.html">RemoveSpuriousRegions()</a></code>

        </dt>
        <dd>Removes specific genomic regions (centromeres and telomeres) of all chromosomes</dd>
      </dl><dl><dt>

          <code><a href="ReplaceQDNAseqAssaySlots.html">ReplaceQDNAseqAssaySlots()</a></code>

        </dt>
        <dd>Replacethe QDNAseq Assay Slots</dd>
      </dl><dl><dt>

          <code><a href="RunShallowHRD.html">RunShallowHRD()</a></code>

        </dt>
        <dd>Run shallowHRD on a dataframe or file containing segmented relative copy-number data.</dd>
      </dl><dl><dt>

          <code><a href="RunShallowHRDFromQDNA.html">RunShallowHRDFromQDNA()</a></code>

        </dt>
        <dd>Run ShallowHRD on a QDNAseq object.</dd>
      </dl><dl><dt>

          <code><a href="RunThresholdSimulations.html">RunThresholdSimulations()</a></code>

        </dt>
        <dd>Runs simulations to obtain a list of possible thresholds from the segment ratio differences</dd>
      </dl><dl><dt>

          <code><a href="SEAlluvialPlot.html">SEAlluvialPlot()</a></code>

        </dt>
        <dd>Make an Alluvial Plot based on maximum signature exposure</dd>
      </dl><dl><dt>

          <code><a href="SegmentsToCopyNumber.html">SegmentsToCopyNumber()</a></code>

        </dt>
        <dd>Transforms segment tables into per-bin copy-number tables.</dd>
      </dl><dl><dt>

          <code><a href="ShrinkReprTMP.html">ShrinkReprTMP()</a></code>

        </dt>
        <dd>(Kind of) merges segments that lie on the same level.</dd>
      </dl><dl><dt>

          <code><a href="SignatureExposuresPlot.html">SignatureExposuresPlot()</a></code>

        </dt>
        <dd>Create Heatmap of Signature Exposures</dd>
      </dl><dl><dt>

          <code><a href="SmallMissingChrArms.html">SmallMissingChrArms()</a></code>

        </dt>
        <dd>Deals with missing chromosome arm values in the given data frame of small segments.</dd>
      </dl><dl><dt>

          <code><a href="SummaryCNPlot.html">SummaryCNPlot()</a></code>

        </dt>
        <dd>Generate Summary Copy-Number Aberrations Plot</dd>
      </dl><dl><dt>

          <code><a href="TumourFraction.html">TumourFraction()</a></code>

        </dt>
        <dd>Compute tumour DNA fraction for the given absolute copy number and cellularity</dd>
      </dl><dl><dt>

          <code><a href="TwoFeatureScatterPlot.html">TwoFeatureScatterPlot()</a></code>

        </dt>
        <dd>Make a two-way scatterplot of two CN-features</dd>
      </dl><dl><dt>

          <code><a href="WassDistancePlot.html">WassDistancePlot()</a></code>

        </dt>
        <dd>Plot a Heatmap of Component-wise Wasserstein Distances</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Maxwell Douglas, Branden Lynch, Carlos Vasquez, Jacky Yiu, Matthew Eldridge, Geoff Macintyre, Ding Ma, Nirupama Tamvada, Sameer Shankar.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

