<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Makes Relative Copy-Number Gain/Loss Summary Tables — MakeSummaryTable • utanos</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Makes Relative Copy-Number Gain/Loss Summary Tables — MakeSummaryTable"><meta name="description" content="MakeSummaryTable takes in a QDNAseq CN-object and returns tables containing agglomerated RCN gain/loss information across the provided samples.
Prior to summarizing the gains/losses, the function applies a log2 transformation to center the data at 0.
Optionally it also writes out a TSV table containing the summary."><meta property="og:description" content="MakeSummaryTable takes in a QDNAseq CN-object and returns tables containing agglomerated RCN gain/loss information across the provided samples.
Prior to summarizing the gains/losses, the function applies a log2 transformation to center the data at 0.
Optionally it also writes out a TSV table containing the summary."><meta property="og:image" content="https://huntsmanlab.github.io/utanos/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">utanos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/utanos.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/A-samplequality-vignette.html">1. Sample Filtering and Quality Evaluation</a></li>
    <li><a class="dropdown-item" href="../articles/B-cndiversityandscaling-vignette.html">2. Investigating Copy-Number Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/C-cnsignatures-vignette.html">3. Investigating and Creating Copy-Number Signatures</a></li>
    <li><a class="dropdown-item" href="../articles/D-shallowhrd-vignette.html">4. Predicting Homologous Recombination Deficiency (HRD)</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Makes Relative Copy-Number Gain/Loss Summary Tables</h1>

      <div class="d-none name"><code>MakeSummaryTable.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>MakeSummaryTable takes in a QDNAseq CN-object and returns tables containing agglomerated RCN gain/loss information across the provided samples. <br>
Prior to summarizing the gains/losses, the function applies a log2 transformation to center the data at 0. <br>
Optionally it also writes out a TSV table containing the summary.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">MakeSummaryTable</span><span class="op">(</span></span>
<span>  <span class="va">CNobj</span>,</span>
<span>  <span class="va">lowT</span>,</span>
<span>  <span class="va">highT</span>,</span>
<span>  pL <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  pG <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  prop <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  find_peaks <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  snames <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ref_genome <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>  save_path <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-cnobj">CNobj<a class="anchor" aria-label="anchor" href="#arg-cnobj"></a></dt>
<dd><p>An S4 object of type QDNAseqCopyNumbers. This object must contain slots for: <br>
copynumber, segmented (standard CN slots) <br>
probgain, probdloss, probamp, probnorm, calls, probloss (from running CGHcall)</p></dd>


<dt id="arg-lowt">lowT<a class="anchor" aria-label="anchor" href="#arg-lowt"></a></dt>
<dd><p>Numeric (negative). Ignore gains/losses between this value and 0. ex. -0.1 <br>
This parameter helps filter out noise.</p></dd>


<dt id="arg-hight">highT<a class="anchor" aria-label="anchor" href="#arg-hight"></a></dt>
<dd><p>Numeric (positive). Ignore gains/losses between this value and 0. ex. 0.1 <br>
This parameter helps filter out noise.</p></dd>


<dt id="arg-pl">pL<a class="anchor" aria-label="anchor" href="#arg-pl"></a></dt>
<dd><p>(optional) Numeric. A probability threshold (between 0 and 1). Mask loss events where the probability does not rise to this threshold.</p></dd>


<dt id="arg-pg">pG<a class="anchor" aria-label="anchor" href="#arg-pg"></a></dt>
<dd><p>(optional) Numeric. A probability threshold (between 0 and 1). Mask gain events where the probability does not rise to this threshold.</p></dd>


<dt id="arg-prop">prop<a class="anchor" aria-label="anchor" href="#arg-prop"></a></dt>
<dd><p>(optional) Numeric. A proportion threshold (between 0 and 1). Exclude gain/loss events that occur in less than this proportion of samples.</p></dd>


<dt id="arg-find-peaks">find_peaks<a class="anchor" aria-label="anchor" href="#arg-find-peaks"></a></dt>
<dd><p>(optional) Logical. If TRUE, returns a table of peaks found in the summarized data.</p></dd>


<dt id="arg-snames">snames<a class="anchor" aria-label="anchor" href="#arg-snames"></a></dt>
<dd><p>(optional) Character vector. A vector of sample IDs that are of interest.</p></dd>


<dt id="arg-ref-genome">ref_genome<a class="anchor" aria-label="anchor" href="#arg-ref-genome"></a></dt>
<dd><p>(optional) String. The reference genome used for alignment. <br>
Options: 'hg19', 'hg38', and 'mm10'</p></dd>


<dt id="arg-save-path">save_path<a class="anchor" aria-label="anchor" href="#arg-save-path"></a></dt>
<dd><p>(optional) String. A path (directory) to where segment tables should be saved. ex. '~/Documents/test_project'.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list of data frames. Always the summary_table, optionally a peaks data frame.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Maxwell Douglas, Branden Lynch, Carlos Vasquez, Jacky Yiu, Matthew Eldridge, Geoff Macintyre, Ding Ma, Nirupama Tamvada, Sameer Shankar.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

