<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>0. utanos quickstart guide • utanos</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="0. utanos quickstart guide">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">utanos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/utanos.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/0_quickstart-vignette.html">0. utanos quickstart guide</a></li>
    <li><a class="dropdown-item" href="../articles/1_samplequality-vignette.html">1. Sample Filtering and Quality Evaluation</a></li>
    <li><a class="dropdown-item" href="../articles/2_cndiversityandscaling-vignette.html">2. Investigating Copy-Number Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/3_cnsignatures-vignette.html">3. Investigating and Creating Copy-Number Signatures</a></li>
    <li><a class="dropdown-item" href="../articles/4_shallowhrd-vignette.html">4. Predicting Homologous Recombination Deficiency (HRD)</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>0. utanos quickstart guide</h1>
            
      

      <div class="d-none name"><code>0_quickstart-vignette.Rmd</code></div>
    </div>

    
    
<p>The goal of the “CN-Signatures Utils” package is to be a swiss army
knife in analyzing shallow/low-pass whole genome sequencing data.
Specifically, the focus of the package is to be on functions to
re-structure data, with a focus on effective visualization and analysis
of CN signatures.</p>
<p>The flowchart represents the steps involved in the analysis of
shallow/low-pass whole genome sequencing data for copy number
signatures. In this vignette, we specifically focus on the relative and
absolute copy-number calling. Generally in such an analysis,
<code>QDNAseq</code> has been commonly used in calling relative
copy-numbers. The output generated by <code>QDNAseq</code> is in the
form of an S4 object.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ccagc/QDNAseq" class="external-link">QDNAseq</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rascal</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="co"># Reading in test data</span></span>
<span><span class="va">qdnaseq_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"~/projects/vignettes/30kb_rCN_comCNVfilt.rds"</span><span class="op">)</span></span>
<span><span class="co"># QDNAseq object</span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">qdnaseq_output</span><span class="op">)</span></span></code></pre></div>
<p>Packages such as <code>QDNAseq</code> are used for summing the number
of reads that align within genomic windows or bins, typically 30kb in
size, and correcting for GC-content and mappability. Thus, these
packages output copy number values that are relative to the average copy
number within the sample for the GC and mappability of each bin.
Relative copy numbers are smoothed and segmented to determine genomic
regions with gains or losses.</p>
<p>However, for inferring copy number signatures, relative copy numbers
must be converted to absolute copy numbers. Relative copy number
profiles can thus be fit to absolute copy numbers by evaluating various
estimates of ploidy and cellularity, and selecting the best fit
estimates. In the workflow native to the CN-Sigs utils package, the R
package <code>rascal</code> is used to call absolute copy numbers.</p>
<p>The batch fitting script for <code>rascal</code> can be run on the
command line to estimate tumour ploidies and cellularities for multiple
samples. The function that performs this in the <code>rascal</code>
package is <code>copy_number_for_sample.R</code>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>Rscript fit_absolute_copy_numbers.R <span class="sc">-</span>i <span class="dv">30</span>kb_rCN_comCNVfilt.rds <span class="sc">-</span>o <span class="dv">30</span>kb_aCNsolutions.csv</span></code></pre></div>
<p><code>rascal</code> accepts either a RDS file containing a
<code>QDNAseqCopyNumbers</code> object/ a dataframe/RDS file containing
the following columns:</p>
<ul>
<li>sample</li>
<li>chromosome</li>
<li>start</li>
<li>end</li>
<li>segmented</li>
</ul>
<p>The output format can be chosen by the user. In this case, running
<code>rascal</code> outputs a .csv file of all the solutions found all
the ploidy and cellularity solutions found for a particular sample.
<code>rascal</code>’s workflow includes the mean absolute distance (MAD)
(which is the default) or root mean square deviation (RMSD) to find the
optimal ploidy and cellularity solutions.</p>
<div class="section level2">
<h2 id="calculate-acns">Calculate ACNs<a class="anchor" aria-label="anchor" href="#calculate-acns"></a>
</h2>
<p>The CalculateACNs function is extremely important, as it calculates
the absolute copy numbers (ACNs) from the relative copy number profiles
of one or more samples. In order to understand what this function does,
it is important to begin by clarifying the inputs that the function
expects.</p>
<p>The first argument expected by CalculateACNs, relative_segs, is a
tab-separated values (TSV) file containing the relative segmented
copy-numbers.</p>
<p>The table above shows the feature (in the following format,
chromosome:start-end), chromosome, start and end, followed by individual
sample relative segmented copy-numbers (wide format data).</p>
<p>The second argument, acnmethod, is a string which represents the
method by which ACNs are being calculated. It can be one of: - “maxvaf”,
- a character vector of the genes from which to pull VAFs, where genes
are assumed to be in order of decreasing precedence, ex. c(‘TP53’,
‘KRAS’, ‘PTEN’), - “mad”</p>
<p>The third argument, rascal_sols, is a CSV file of the calculated
rascal solutions.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">solutions</span> <span class="op">&lt;-</span> <span class="st">"~/projects/vignettes/30kb_comCNVfilt_rascal_solutions.csv"</span></span>
<span></span>
<span><span class="va">solutions</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">solutions</span>, sep <span class="op">=</span> <span class="st">','</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">solutions</span><span class="op">)</span></span></code></pre></div>
<p>The table above has the ploidy, cellularity and distance columns. The
distance is calculated using a rascal helper function (based on mean
absolute difference (MAD) or root mean square different (RMSD)).</p>
<p>The fourth argument for the CalculateACNs, variants, is the variants
table, which is a comma-separated values (CSV) file containing
long-format data. The data appear as follows:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants_path</span> <span class="op">&lt;-</span> <span class="st">"~/projects/vignettes/p53abn_endometrial_variants_customoutput_20230413.csv"</span></span>
<span></span>
<span><span class="va">variants</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">variants_path</span>, sep <span class="op">=</span> <span class="st">','</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#rename gene_symbol to gene_name in order to meet the input requirements</span></span>
<span><span class="va">variants</span> <span class="op">&lt;-</span> <span class="va">variants</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>gene_name <span class="op">=</span> <span class="va">gene_symbol</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">variants</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>The variants table is a dataframe of the variants including variant
allele frequencies per gene and per sample. The following are the
requirements for the variants table when using VAFs from targeted panel
sequencing or some other technology:</p>
<ul>
<li>The variants must must be in a datatable/dataframe</li>
<li>The required columns are as follows: sample_id, chromosome, start,
end, gene_name, ref, alt, vaf (with these exact names, but in any
order)</li>
<li>Each row of said table must correspond to a unique variant</li>
<li>Each variant must have an associated variant allele frequency</li>
<li>Each row must also be associated with a specific sample</li>
</ul>
<p>Note that this example variants table has more variant information
than is absolutely necessary for the input.</p>
<p>The fifth argument is relative_cns; a TSV of the relative
copy-numbers. It is mandatory if ‘addplots’ param (a parameter that
allows plots to be returned alongisde the ACNs) is set to true.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rel_cns</span> <span class="op">&lt;-</span> <span class="st">"~/projects/vignettes/30kb_rCN_cns_comCNVfilt.tsv"</span></span>
<span></span>
<span><span class="va">rel_cns</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">rel_cns</span>, sep <span class="op">=</span> <span class="st">'\t'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">rel_cns</span><span class="op">[</span> ,<span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>The table above shows the feature (in the following format,
chromosome:start-end), chromosome, start and end, followed by individual
sample relative copy-numbers (wide format data).</p>
<p>Please note, the first argument (relative_segs) and fifth argument
(relative_cns) seem very similar in terms of what the tables look like,
but are quite different; relative_segs is a TSV containing the relative
segmented copy-numbers, whereas relative_cns is a TSV containing the
relative copy-numbers.</p>
<p>In addition to the five parameters, there are three optional
parameters;</p>
<ul>
<li>addplots, Logical. This indicates whether or not plots should be
returned alongside the ACNs.</li>
<li>acn_save_path, String. The output path where the result will be
saved. An absolute path is recommended for this, ex.
‘/Users/newuser/Documents/output.rds’</li>
<li>return_sols, Logical. Returns the selected rascal solution.</li>
</ul>
<p>Lastly, here is what the output of CalculateACNs appears to be;</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example of output file</span></span>
<span><span class="va">output</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"~/projects/vignettes/30kb_aCN_rascal_segments_comCNVfilt.rds"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">output</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>CalculateACNs returns an R Data Serialization (RDS) file. This output
file is a list of named dataframes, where each dataframe corresponds to
a different sample where an Absolute Copy Number profile was
successfully found. Optionally a second list of plot objects (ggplot);
one for each sample where a profile was found.</p>
</div>
<div class="section level2">
<h2 id="what-features-is-this-workflow-missing">What Features is this Workflow missing?<a class="anchor" aria-label="anchor" href="#what-features-is-this-workflow-missing"></a>
</h2>
<p>This vignette contains the initial workflow for absolute copy number
calling that is an important part of the pipeline in inferring copy
number signatures. Some features missing in this workflow that would be
useful additions to the package are the following:</p>
<p>Functions to evaluate the quality of the samples are likely to be a
crucial part of this package at all stages of copy number and signature
calling. In particular this includes:</p>
<ul>
<li>A function to output metrics to evaluate the quality of the
<code>QDNASeq</code> relative copy numbers, as well as the
<code>rascal</code> absolute copy numbers. It is particularly important
to have a designated metric assigned to evaluating the
<code>rascal</code> copy number calls of multiple samples. While
<code>rascal</code>’s workflow includes the mean absolute distance (MAD)
or root mean square deviation (RMSD) to find the optimal ploidy and
cellularity solutions, the <code>rascal</code> authors further recommend
manually clicking through segment copy number plots of multiple samples
to evaluate the quality of th calls manually. This is a tedious process
that should be automated for better usability of the copy number calling
pipeline. Some metrics and additions to think about are the
following:</li>
<li>
<a href="https://academic.oup.com/nar/article/47/4/1605/5253050?login=true" class="external-link">Metrics
used to compare QDNAseq and WisecondorX</a>. In particular, the observed
median segment variance (MSVo) seems interesting but there are other
statistics in this paper that should likely be implemented in the
package such as the Lilliefors normality statistic. The MVSo, a
sample-wise measure for noise, is defined as the median of a set of
variances, where each variance corresponds to the variance of a segment.
This metric is inversely proportional to read depth and selected bin
size, and reflects the noise as a sample-wide measure.</li>
<li>
<a href="https://static-content.springer.com/esm/art%3A10.1186%2Fs13073-020-00735-4/MediaObjects/13073_2020_735_MOESM2_ESM.docx" class="external-link">The
Copy number profile abnormality score</a>. The CPA score is designed to
distinguish abnormal from healthy copy number profiles and contains a
penalty term for shorter segments, which are exclusively observed in bad
quality or truly highly aberrant samples.</li>
</ul>
<p>Some ideas of features to implement are the following:</p>
<ol style="list-style-type: decimal">
<li><p>Function to output a table of these statistics per sample. We
should test out this statistics on our data by looking at how well they
capture what we see in the manual evaluations for some of the samples,
and make a vignette of a recommended workflow to use for maximum
usability of the package.</p></li>
<li><p>Plots which capture the distribution of these statistics for the
entire sample set such as a line plot or histogram.</p></li>
<li><p>It would be interesting to capture the “outliers” relative to the
other samples (an apples-to-apples comparison of tumour samples to the
others). Thus a function that could automate this and output the copy
number profile plots for just a couple of these samples with abberant
copy number profiles would be very useful.</p></li>
</ol>
</div>
<div class="section level2">
<h2 id="general-package-functionalities-to-work-on">General Package Functionalities to work on<a class="anchor" aria-label="anchor" href="#general-package-functionalities-to-work-on"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Re-factoring and cleaning code</li>
<li>Documentation</li>
<li>Flexible QDNASeq S4 object</li>
<li>Consistent input and output formats</li>
<li>Unit tests</li>
</ol>
</div>
<div class="section level2">
<h2 id="what-comes-next-in-inferring-cn-signatures">What comes next in inferring CN-signatures?<a class="anchor" aria-label="anchor" href="#what-comes-next-in-inferring-cn-signatures"></a>
</h2>
<p>The following downstream analyses follow after the absolute copy
number calling step in inferring copy number signatures. The package
currently contains functions for these which need to be modified
next:</p>
<ol style="list-style-type: decimal">
<li>Inferring the distributions of absolute copy number feature
distributions - the breakpoint count per 10Mb, the copy number of the
segments, the difference in copy number between adjacent segments, the
breakpoint count per chromosome arm, the lengths of oscillating copy
number segment chains and the size of segments. A mixture model is then
fit to these six distributions, This directly corresponds to the figure
below <a href="https://www.nature.com/articles/s41588-018-0179-8.pdf" class="external-link">(Macintyre
et al., 2018)</a>.</li>
</ol>
<p><img src="images/distc.png" width="1058"></p>
<ol start="2" style="list-style-type: decimal">
<li>Apply non-megative matrix factorization to visualize the
tumor-by-signature matrix, the signature-by-component matrix and copy
numbers by the copy number features which is an important end-result of
this package. Ideally, we want to add an intelligent plotting function
that is capable of plotting these visualizations side-by-side with
annotations for metadata, etc.</li>
</ol>
<p><img src="images/papf.png" width="757"></p>
<p>(from <a href="https://www.nature.com/articles/s41588-018-0179-8.pdf" class="external-link">(Macintyre
et al., 2018)</a>)</p>
<div class="section level3">
<h3 id="ideal-package-output">Ideal Package Output<a class="anchor" aria-label="anchor" href="#ideal-package-output"></a>
</h3>
<p><img src="images/cronfig.png" width="836"></p>
<p><strong>Note:</strong> Sample quality evaluation at this stage is a
key need this package aims to meet. Currently there are no well
established methods to evaluate the quality of these signature
calls.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Maxwell Douglas, Branden Lynch, Carlos Vasquez, Jacky Yiu, Ding Ma, Nirupama Tamvada, Sameer Shankar.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
