<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>1. Sample Filtering and Quality Evaluation • utanos</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="1. Sample Filtering and Quality Evaluation">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">utanos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/utanos.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/0_quickstart-vignette.html">0. utanos quickstart guide</a></li>
    <li><a class="dropdown-item" href="../articles/1_samplequality-vignette.html">1. Sample Filtering and Quality Evaluation</a></li>
    <li><a class="dropdown-item" href="../articles/2_cndiversityandscaling-vignette.html">2. Investigating Copy-Number Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/3_cnsignatures-vignette.html">3. Investigating and Creating Copy-Number Signatures</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>1. Sample Filtering and Quality Evaluation</h1>
            
      

      <div class="d-none name"><code>1_samplequality-vignette.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The utanos package begins with relative copy-number (rCN) call
genomic data. It expects rCN data output from tools such as
<code>QDNASeq</code> or <code>WiseCondorX</code>. This data consists of
calls (a numeric value) for each ‘bin’ in the genome. A bin is an
artificially defined partitioning of the genome into pieces of
sufficient size to reasonably estimate a relative copy-number. For
shallow sequencing this is needed because there is not sufficient depth
present for base-pair level resolution.</p>
<p>Tabular long-format example:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>chromosome   start     end sample    CN.call</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="sc">&lt;</span>chr<span class="sc">&gt;</span>        <span class="er">&lt;</span>int<span class="sc">&gt;</span>   <span class="er">&lt;</span>int<span class="sc">&gt;</span> <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       <span class="er">&lt;</span>dbl<span class="sc">&gt;</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="dv">1</span>           <span class="dv">900001</span>  <span class="dv">950000</span> VOA14948A    <span class="fl">1.00</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="dv">1</span>           <span class="dv">950001</span> <span class="dv">1000000</span> VOA14948A    <span class="fl">1.00</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="dv">1</span>          <span class="dv">1000001</span> <span class="dv">1050000</span> VOA14948A    <span class="fl">1.00</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="dv">1</span>          <span class="dv">1050001</span> <span class="dv">1100000</span> VOA14948A    <span class="fl">1.00</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="dv">1</span>          <span class="dv">1100001</span> <span class="dv">1150000</span> VOA14948A    <span class="fl">0.96</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="dv">1</span>          <span class="dv">1150001</span> <span class="dv">1200000</span> VOA14948A    <span class="fl">0.96</span></span></code></pre></div>
<p>Relative CN-callers such as <code>QDNAseq</code> sometimes have
built-in QC functionality for filtering out bins. Most commonly this is
for low-mappability or blacklist regions (<a href="https://doi.org/10.1101%2Fgr.175141.114" class="external-link">see the paper for
descriptions</a>). Other CN-callers also can have whole-sample QC
evaluation options though the vast majority of these tend to be very
manual. They require the researcher to examine every individual sample.
This vignette will cover how utanos can be used to do filtering and QC
evaluation on an individual or group level.</p>
</div>
<div class="section level2">
<h2 id="i---filtration">I - Filtration<a class="anchor" aria-label="anchor" href="#i---filtration"></a>
</h2>
<p>Often, in order to focus on the CN-aberrations of most interest, it
is convenient to filter out more than just blacklist regions. What about
centromeres? Telomeres? There are regions of the genome that are
commonly copy-number aberrant in the background human population.
Perhaps it would be nice to mask these too. Utanos clarifies the
filtration process.</p>
<p>Note: Utanos comes with 2 tables of common variants (one for hg19 and
the other hg38). They can be used to filter a dataset like so.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rcn.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"~/Path_to_your_data/a_data_object_aligned_to_hg38.rds"</span><span class="op">)</span></span>
<span><span class="va">rcn.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FilterCNs.html">FilterCNs</a></span><span class="op">(</span>cnobj <span class="op">=</span> <span class="va">rcn.obj</span>,</span>
<span>                     genome <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>                     minimum_overlap <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>                     removebins <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                     maskgaps <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     maskcomCNVs <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><a href="http://dgv.tcag.ca/dgv/app/downloads?ref=" class="external-link">Database of
Genomic Variants (DGV)</a><br><a href="https://pubmed.ncbi.nlm.nih.gov/24174537/" class="external-link">The associated
paper</a></p>
<p>But let’s break down this filtration step using a real example
dataset. An accompanying data package to utanos is
<em>utanosmodellingdata</em> found <a href="https://github.com/Huntsmanlab/utanosmodellingdata" class="external-link">here</a>.
Clone that repo to somewhere convenient such as a common ‘repos’ folder
on your machine and read in the sample data. It is human endometrial
carcinoma sWGS data aligned to hg19.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(utanos)</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(magrittr)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="sc">&gt;</span> rcn.obj <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"~/repos/utanosmodellingdata/sample_copynumber_data/sample_rcn_data.rds"</span>)</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="sc">&gt;</span> rcn.obj <span class="ot">&lt;-</span> <span class="fu">FilterCNs</span>(<span class="at">cnobj =</span> rcn.obj,</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>                       <span class="at">genome =</span> <span class="st">"hg19"</span>,</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>                       <span class="at">minimum_overlap =</span> <span class="dv">5000</span>,</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>                       <span class="at">removebins =</span> <span class="cn">FALSE</span>,</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>                       <span class="at">maskgaps =</span> <span class="cn">TRUE</span>,</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>                       <span class="at">maskcomCNVs =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">featureData</span>(rcn.obj)<span class="sc">@</span>data <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">5</span>)</span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>                chromosome  start    end     bases       gc mappability blacklist residual   use</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span><span class="dv">1-30000</span>                <span class="dv">1</span>      <span class="dv">1</span>  <span class="dv">30000</span>  <span class="fl">66.66667</span> <span class="fl">55.48500</span>    <span class="fl">12.39630</span>  <span class="fl">3.263333</span>      <span class="cn">NaN</span> <span class="cn">FALSE</span></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span><span class="dv">30001-60000</span>            <span class="dv">1</span>  <span class="dv">30001</span>  <span class="dv">60000</span> <span class="fl">100.00000</span> <span class="fl">39.75667</span>    <span class="fl">28.22070</span>  <span class="fl">0.000000</span>      <span class="cn">NaN</span> <span class="cn">FALSE</span></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span><span class="dv">60001-90000</span>            <span class="dv">1</span>  <span class="dv">60001</span>  <span class="dv">90000</span> <span class="fl">100.00000</span> <span class="fl">36.13000</span>    <span class="fl">27.47780</span>  <span class="fl">0.000000</span>      <span class="cn">NaN</span> <span class="cn">FALSE</span></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span><span class="dv">90001-120000</span>           <span class="dv">1</span>  <span class="dv">90001</span> <span class="dv">120000</span> <span class="fl">100.00000</span> <span class="fl">39.61667</span>     <span class="fl">9.05091</span>  <span class="fl">0.000000</span>      <span class="cn">NaN</span> <span class="cn">FALSE</span></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span><span class="dv">120001-150000</span>          <span class="dv">1</span> <span class="dv">120001</span> <span class="dv">150000</span> <span class="fl">100.00000</span> <span class="fl">46.52333</span>     <span class="fl">7.33796</span>  <span class="fl">0.000000</span>      <span class="cn">NaN</span> <span class="cn">FALSE</span></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a>                centro.telo.mask comCNV.mask</span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span><span class="dv">1-30000</span>                  <span class="cn">FALSE</span>       <span class="cn">FALSE</span></span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span><span class="dv">30001-60000</span>               <span class="cn">TRUE</span>       <span class="cn">FALSE</span></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span><span class="dv">60001-90000</span>               <span class="cn">TRUE</span>       <span class="cn">FALSE</span></span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span><span class="dv">90001-120000</span>              <span class="cn">TRUE</span>       <span class="cn">FALSE</span></span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span><span class="dv">120001-150000</span>             <span class="cn">TRUE</span>       <span class="cn">FALSE</span></span></code></pre></div>
<p>This sample data had relative copy-numbers called by QDNAseq. The
‘use’ column in the S4 object indicates bins QDNAseq determines to be
useable. The bins with ‘FALSE’ are not. For more details on QDNAseq and
its methods see the bioconductor vignette <a href="https://bioconductor.org/packages/release/bioc/html/QDNAseq.html" class="external-link">here</a>.
The <code>FilterCNs</code> utanos function adds another two columns to
the featureData table because the <code>maskgaps</code> and
<code>maskcomCNVs</code> parameters were set to TRUE. The
<code>use</code>, <code>centro.telo.mask</code>, and
<code>comCNV.mask</code> columns we will refer to as ‘masks’, as if you
subset by these columns you essentially mask them from downstream
analysis.<br>
A final note on the <code>FilterCNs</code> function: using the
<code>filter_by</code> parameter a user can provide their own regions by
which to create a mask or filter the QDNAseq copy-number object. Just
pass a dataframe with the locations, one location per row, to the
<code>filter_by</code> parameter. Using the <code>maskname</code>
parameter the user can set the name for the mask. The
<code>maskname</code> parameter expects a character string.</p>
<p>Next, we can visualize the masks using the
<code>CNSegmentsPlot</code> function.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="sc">&gt;</span> a <span class="ot">&lt;-</span> <span class="fu">CNSegmentsPlot</span>(rcn.obj,</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>                      <span class="at">sample =</span> <span class="st">"CC-CHM-1341"</span>,</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>                      <span class="at">highlight_masks =</span> <span class="fu">c</span>(<span class="st">'comCNV.mask'</span>),</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>                      <span class="at">point_size =</span> <span class="fl">0.5</span>,</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>                      <span class="at">dolog2 =</span> <span class="cn">TRUE</span>, </span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>                      <span class="at">def_point_colour =</span> <span class="st">'grey'</span>)</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="sc">&gt;</span> a</span></code></pre></div>
<p><img src="images/sampleQC_vignette_01_CC-CHM-1341.png" width="100%"></p>
<p>Clearly visible in the plot is that several of the regions that are
commonly copy-number variant in humans coincide with a
copy-number/segmentation change. Depending on the user’s specific use
case these regions may or may not be worth excluding.<br>
Note: The <code>dolog2</code> parameter was set to TRUE in this call so
we see the copy-number calls and segments log2 transformed and so
centered about 0. See the documentation for that function for more
details.</p>
</div>
<div class="section level2">
<h2 id="ii---sample-quality-evaluation">II - Sample Quality Evaluation<a class="anchor" aria-label="anchor" href="#ii---sample-quality-evaluation"></a>
</h2>
<p>Shallowly sequenced whole genome data involves copy-number calling in
two stages: relative copy-calling (via tools such as
<code>QDNASeq</code> and <code>WiseCondorX</code>) and then absolute
copy number calling which is run on the relative calls (via tools like
<code>rascal</code>). Naturally, the accuracy of the inferences that can
be made in downstream analysis depend heavily on the accuracy of the
copy number calls inferred. However, for large projects it is an
intensive demand to manually evaluate the quality of hundreds to
thousands of samples. This is another the niche that Utanos fills.</p>
<p>Use the <code>GetSampleQualityDecision</code> function to get quality
calls for each sample in your dataset. The <code>QualityPlot</code>
function can then be used to visualize the samples that fall outside the
thresholds.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="sc">&gt;</span> qc_decisions <span class="ot">&lt;-</span> <span class="fu">GetSampleQualityDecision</span>(rcn.obj)</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="sc">&gt;</span> qc_plot <span class="ot">&lt;-</span> <span class="fu">QualityPlot</span>(qc_decisions)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="sc">&gt;</span> qc_decisions <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">5</span>)</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>       sample seg_counts  median_sd decision seg_cut    med_cut</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="dv">1</span> CC<span class="sc">-</span>CHM<span class="dv">-1341</span>        <span class="dv">204</span> <span class="fl">0.07696209</span>     High   <span class="fl">620.3</span> <span class="fl">0.09096386</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="dv">2</span> CC<span class="sc">-</span>CHM<span class="dv">-1347</span>        <span class="dv">695</span> <span class="fl">0.07137276</span>      Low   <span class="fl">620.3</span> <span class="fl">0.09096386</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="dv">3</span> CC<span class="sc">-</span>CHM<span class="dv">-1353</span>        <span class="dv">235</span> <span class="fl">0.06784682</span>     High   <span class="fl">620.3</span> <span class="fl">0.09096386</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="dv">4</span> CC<span class="sc">-</span>CHM<span class="dv">-1355</span>        <span class="dv">240</span> <span class="fl">0.07509611</span>     High   <span class="fl">620.3</span> <span class="fl">0.09096386</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="dv">5</span> CC<span class="sc">-</span>CHM<span class="dv">-1361</span>        <span class="dv">529</span> <span class="fl">0.07547814</span>     High   <span class="fl">620.3</span> <span class="fl">0.09096386</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="sc">&gt;</span> qc_plot</span></code></pre></div>
<p><img src="images/sampleQC_vignette_02_qcdecisions.png" width="60%"></p>
<p>The default behaviour of this function is make a sample-wise decision
based on thresholds for the median segment standard deviation and the
total segment count.</p>
<div class="section level3">
<h3 id="ii-1---about-the-metrics-median-segment-variance-and-segment-count">II.1 - About the metrics: Median segment variance and segment
count<a class="anchor" aria-label="anchor" href="#ii-1---about-the-metrics-median-segment-variance-and-segment-count"></a>
</h3>
<p>The segment variance is calculated as the sample variance per
segment, i.e the variance between the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>l</mi><mi>o</mi><msub><mi>g</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">log_2</annotation></semantics></math>
ratios of each bin in each segment. The observed median segment variance
then is a sample-wise measure for noise, which is defined as the median
of a set of variances.</p>
<p>The segment count is simply the number of distinct segments found in
the sample. In poor quality samples over-segementation is a frequent
indicator.</p>
<p>The strategy with this quality evaluation technique is to consider
both of these metrics in tandem. To work through the intuition, consider
the following two copy-number profiles:</p>
<div class="section level4">
<h4 id="sample-1-good-quality-copy-number-profile">Sample 1: Good quality copy-number profile<a class="anchor" aria-label="anchor" href="#sample-1-good-quality-copy-number-profile"></a>
</h4>
<p><img src="images/sampleQC_vignette_03_CC-CHM-1341.png" width="100%"></p>
</div>
<div class="section level4">
<h4 id="sample-2-poorer-quality-copy-number-profile">Sample 2: Poorer quality copy number profile<a class="anchor" aria-label="anchor" href="#sample-2-poorer-quality-copy-number-profile"></a>
</h4>
<p><img src="images/sampleQC_vignette_04_VS09-16339-B8.png" width="100%"></p>
<p>The key aspects that differ between the two profiles are: the number
of segments, the length of segments, and the variance within each
segment of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>l</mi><mi>o</mi><msub><mi>g</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">log_2</annotation></semantics></math>
ratios, i.e we see the copy number ratios more tightly clustered around
the orange line for sample 1 compared to sample 2.</p>
<p><br><br><br></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="iii---possible-future-additions-and-ideas-the-copy-number-profile-abnormality-score-penalty">III - Possible future additions and ideas: The copy-number profile
abnormality score penalty<a class="anchor" aria-label="anchor" href="#iii---possible-future-additions-and-ideas-the-copy-number-profile-abnormality-score-penalty"></a>
</h2>
<p>This is a metric to distinguish abnormal from healthy copy number
profiles. The score quantifies the deviation of segments from the normal
diploid state at the sample-level. The copy-number profile abnormality
score can be expressed as follows:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><mi>P</mi><mi>A</mi><mspace width="0.222em"></mspace><mo>=</mo><mspace width="0.222em"></mspace><mfrac><mrow><msubsup><mi>Σ</mi><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></msubsup><mrow><mo stretchy="true" form="prefix">(</mo><mrow><mo stretchy="true" form="prefix">|</mo><msub><mi>Z</mi><mrow><mi>s</mi><mi>e</mi><mi>g</mi><mi>m</mi><mi>e</mi><mi>n</mi><msub><mi>t</mi><mi>i</mi></msub></mrow></msub><mo stretchy="true" form="postfix">|</mo></mrow><mo>*</mo><msub><mi>l</mi><mrow><mi>s</mi><mi>e</mi><mi>g</mi><mi>m</mi><mi>e</mi><mi>n</mi><msub><mi>t</mi><mi>i</mi></msub></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mi>n</mi></mfrac></mrow><annotation encoding="application/x-tex">CPA \ = \ \frac{\Sigma^{n}_{i=1} (|Z_{segment_i}|* l_{segment_i})}{n}</annotation></semantics></math></p>
<p>Here,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Z</mi><mrow><mi>s</mi><mi>e</mi><mi>g</mi><mi>m</mi><mi>e</mi><mi>n</mi><msub><mi>t</mi><mi>i</mi></msub></mrow></msub><annotation encoding="application/x-tex">Z_{segment_i}</annotation></semantics></math>
represents the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Z</mi><annotation encoding="application/x-tex">Z</annotation></semantics></math>-score
of the segment
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>l</mi><mrow><mi>s</mi><mi>e</mi><mi>g</mi><mi>m</mi><mi>e</mi><mi>n</mi><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">l_{segment}</annotation></semantics></math>
represents the length of the segment and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
represents the number of segments.<br>
The average segment length can be represented as follows:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><mrow><msubsup><mi>Σ</mi><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></msubsup><msub><mi>l</mi><mrow><mi>s</mi><mi>e</mi><mi>g</mi><mi>m</mi><mi>e</mi><mi>n</mi><msub><mi>t</mi><mi>i</mi></msub></mrow></msub></mrow><mi>n</mi></mfrac><annotation encoding="application/x-tex">\frac{\Sigma^{n}_{i=1}l_{segment_i}}{n}</annotation></semantics></math></p>
<p>This term functions as a penalty term for sample quality, as short
segments are generally observed in bad quality or truly highly aberrant
samples. Thus, calculating this penalty in addition to other metrics
could be very informative. A very important note is that alone this
metric would be problematic. Copy-number aberrant is an expected state
when the biology of the sample truly is chromosomal aberrations. This
metric in tandem with others would seek to find aberration where none
should be present.</p>
<p>Thus, a future workflow could be a classifier (which would be trained
on our test set) with the features being the segment sizes, median
segment variances, the Lilliefor normality statistic, read length,
coverage etc.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Maxwell Douglas, Branden Lynch, Carlos Vasquez, Jacky Yiu, Ding Ma, Nirupama Tamvada, Sameer Shankar.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
