<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>2. Investigating Copy-Number Diversity • utanos</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="2. Investigating Copy-Number Diversity">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">utanos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/utanos.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/0_quickstart-vignette.html">0. utanos quickstart guide</a></li>
    <li><a class="dropdown-item" href="../articles/1_samplequality-vignette.html">1. Sample Filtering and Quality Evaluation</a></li>
    <li><a class="dropdown-item" href="../articles/2_cndiversityandscaling-vignette.html">2. Investigating Copy-Number Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/3_cnsignatures-vignette.html">3. Investigating and Creating Copy-Number Signatures</a></li>
    <li><a class="dropdown-item" href="../articles/4_shallowhrd-vignette.html">4. Predicting Homologous Recombination Deficiency (HRD)</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>2. Investigating Copy-Number Diversity</h1>
            
      

      <div class="d-none name"><code>2_cndiversityandscaling-vignette.Rmd</code></div>
    </div>

    
    
<hr>
<div class="section level2">
<h2 id="i---introduction">I - Introduction<a class="anchor" aria-label="anchor" href="#i---introduction"></a>
</h2>
<p>Broadly, this vignette seeks to demonstrate how a user could answer
the following three questions using <strong>Utanos</strong>:</p>
<ul>
<li>In my dataset, where do copy-number (CN) gains and losses
co-occur?</li>
<li>To what extent do these aberrations co-occur across individual
samples?</li>
<li>How can this be done at the relative-CN vs absolute-CN scales?</li>
</ul>
<p><strong>Utanos</strong> provides two options by which to find answers
to these questions.<br>
1 - using plots/diagrams to visualize the differences, or 2 - by
manually examining informative, well-formatted tables.</p>
<p><br><strong>Exclusions:</strong> This vignette does not cover data
filtration or setting masks in the QDNAseq object in any detail. The
vignette titled “Sample Filtering and Quality Evaluation” covers the
filtration and QC steps. <br><br></p>
<p>As an example dataset we will again use the samples found in the
<em>utanosmodellingdata</em> repository found <a href="https://github.com/Huntsmanlab/utanosmodellingdata" class="external-link">here</a>. If
not already done, clone that repo to somewhere convenient such as a
common ‘repos’ folder on your machine and read in the example data. It
is human endometrial carcinoma sWGS data aligned to hg19. The associated
publication (Jamieson, 2024) was published in CCR and can be found <a href="https://doi.org/10.1158/1078-0432.CCR-23-3689" class="external-link">here</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(utanos)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(QDNAseq)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(magrittr)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="sc">&gt;</span> rcn.obj <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"~/repos/utanosmodellingdata/sample_copynumber_data/sample_rcn_data.rds"</span>)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="sc">&gt;</span> rcn.obj</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">QDNAseqCopyNumbers</span> (storageMode<span class="sc">:</span> lockedEnvironment)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>assayData<span class="sc">:</span> <span class="dv">103199</span> features, <span class="dv">10</span> samples </span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>  element names<span class="sc">:</span> copynumber, segmented </span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>protocolData<span class="sc">:</span> none</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>phenoData</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>  sampleNames<span class="sc">:</span> CC<span class="sc">-</span>CHM<span class="dv">-1341</span> CC<span class="sc">-</span>CHM<span class="dv">-1347</span> ... CC<span class="sc">-</span>HAM<span class="dv">-0385</span> (<span class="dv">10</span> total)</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a>  varLabels<span class="sc">:</span> name total.reads ... <span class="fu">loess.family</span> (<span class="dv">6</span> total)</span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>  varMetadata<span class="sc">:</span> labelDescription</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>featureData</span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a>  featureNames<span class="sc">:</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">1-30000</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30001-60000</span> ... Y<span class="sc">:</span><span class="dv">59370001-59373566</span> (<span class="dv">103199</span> total)</span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a>  fvarLabels<span class="sc">:</span> chromosome start ... <span class="fu">comCNV.mask</span> (<span class="dv">11</span> total)</span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a>  fvarMetadata<span class="sc">:</span> labelDescription</span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a>experimentData<span class="sc">:</span> use <span class="st">'experimentData(object)'</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a>Annotation<span class="sc">:</span> </span></code></pre></div>
</div>
<div class="section level2">
<h2 id="ii---call-gains-and-losses-on-relative-copy-number-data">II - Call gains and losses on relative copy-number data<a class="anchor" aria-label="anchor" href="#ii---call-gains-and-losses-on-relative-copy-number-data"></a>
</h2>
<p>The callBins function from the QDNAseq package can be run on
segmented copy number data to determine chromosomal aberrations. This
function adds gain/loss probability slots to the input S4 object. Under
the hood it makes use of the CGHcall R package. Please refer to those
packages for further algorithmic details.</p>
<p>Here is an example after first filtering as outlined in the previous
vignette.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="sc">&gt;</span> rcn.obj <span class="ot">&lt;-</span> <span class="fu">FilterCNs</span>(<span class="at">cnobj =</span> rcn.obj, <span class="at">genome =</span> <span class="st">"hg19"</span>, </span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>                       <span class="at">maskcomCNVs =</span> <span class="cn">TRUE</span>, <span class="at">removebins =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="sc">&gt;</span> rcn.obj <span class="ot">&lt;-</span> <span class="fu">callBins</span>(rcn.obj)</span></code></pre></div>
<p>Notes:</p>
<ul>
<li>Depending on dataset size the <code><a href="https://rdrr.io/pkg/QDNAseq/man/callBins.html" class="external-link">callBins()</a></code> function can
take a while</li>
</ul>
</div>
<div class="section level2">
<h2 id="iii---plotting">III - Plotting<a class="anchor" aria-label="anchor" href="#iii---plotting"></a>
</h2>
<div class="section level3">
<h3 id="iii-1---cn-summary-plot">III.1 - CN-Summary Plot<a class="anchor" aria-label="anchor" href="#iii-1---cn-summary-plot"></a>
</h3>
<p>The CGHbase R-package provides an excellent plot showing the gain and
loss probability across multiple samples. <strong>Utanos</strong>
contains a modified version of this plotting function and it can be
generated like so…</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">SummaryCNPlot</span>(rcn.obj, <span class="at">maskaberr =</span> <span class="fl">0.1</span>, <span class="at">maskprob =</span> <span class="fl">0.2</span>)</span></code></pre></div>
<p><img src="images/cndiversity_vignette_01.png" width="100%"></p>
<p>Notes:</p>
<ul>
<li>The ‘maskaberr’ parameter - Default Value: 0.1. This parameter
filters out all gains and losses across samples where the CN-change (the
aberration itself) lies between this value and zero.</li>
<li>The ‘maskprob’ parameter - Default Value: 0.2. This parameter
filters out all the gain/loss probabilities below this value. The
callBins function assigns a probability of gain/loss to each bin of each
sample. This is especially useful when analyzing noisy samples.</li>
</ul>
<p>These parameters are quite similar but both are available because it
allows users to mask low probability gain/loss regions without taking
into account how large the gain/loss was.</p>
</div>
<div class="section level3">
<h3 id="iii-2---relative-cn-diversity-plot">III.2 - Relative-CN Diversity Plot<a class="anchor" aria-label="anchor" href="#iii-2---relative-cn-diversity-plot"></a>
</h3>
<p>Alternatively, the relative copy-number changes can be visualized
sample-by-sample as a heatmap with red signifying a relative gain and
blue a loss. For just a few samples, row-wise clustering doesn’t have
too much of an effect on the plot, but for larger cohorts setting the
<code>cluster</code> parameter to <code>TRUE</code> can really help to
highlight trends in the data. This is a simple hierarchical
clustering.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="sc">&gt;</span> rcn.plot <span class="ot">&lt;-</span> <span class="fu">RCNDiversityPlot</span>(rcn.obj, <span class="at">Xchr =</span> <span class="cn">TRUE</span>, <span class="at">cluster =</span> <span class="cn">TRUE</span>)</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="sc">&gt;</span> rcn.plot</span></code></pre></div>
<p><img src="images/cndiversity_vignette_02.png" width="100%"></p>
</div>
</div>
<div class="section level2">
<h2 id="iv---table-output">IV - Table Output<a class="anchor" aria-label="anchor" href="#iv---table-output"></a>
</h2>
<p>An alternative way to understand the data displayed in the plot for
section III.1 is in tabular format. This allows for custom filtering,
juxtaposing actual numbers, and creating dataset-specific plots. The
<code><a href="../reference/MakeSummaryTable.html">MakeSummaryTable()</a></code> function can be used to generate tables
of the copy-number gains and losses.<br>
If desired, the user can also generate a second table containing just
the peaks by setting <code>find_peaks = TRUE</code>. The threshold
parameters outlined in the below code section will apply to both the
generated summary CN aberrations table and the peaks table. Peaks in the
CN probability data are identified using the
<code>gsignal::findpeaks</code> function. Peaks are simply defined as
local maxima. This function itself is a port-over of its namesake
function in the Octave package ‘signal’.<br>
Alongside the copy-number gains and losses reported in these table is a
column for the genes contained in the region. This is a convenience for
checking whether or not some gene-of-interest is contained within a
copy-number aberration. Just the human and mouse reference genomes are
supported at this point (Ex. ‘mm10’).<br>
The <code><a href="../reference/MakeSummaryTable.html">MakeSummaryTable()</a></code> function returns a list with either
1 or 2 tables (type: tibble).<br><br> These tables can be easily saved as csvs, tsvs, or in your
favourite tabular format for perusal in a different user-friendly
application such as excel.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(gsignal)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co"># Probability of gain or loss threshold - i.e. declare the minimum mean probability of loss or gain across samples</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co"># Expl. For region Chr1:2850001-2865000, and probabilities of loss for 10 samples of:</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co"># 0.814 0 0 0 0 0.125 0.975 0 0 0</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="co"># The mean probability of loss would be 0.1914</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="sc">&gt;</span> prob_loss <span class="ot">&lt;-</span> <span class="fl">0.2</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="sc">&gt;</span> prob_gain <span class="ot">&lt;-</span> <span class="fl">0.2</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="co"># Log space mean relative copy number thresholds</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="co"># To keep things consistent with the plot from the last section, lets use 0.1</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="sc">&gt;</span> low_threshold <span class="ot">&lt;-</span> <span class="sc">-</span><span class="fl">0.1</span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="sc">&gt;</span> high_threshold <span class="ot">&lt;-</span> <span class="fl">0.1</span></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a><span class="co"># Presence threshold - i.e. In what minimum proportion of samples does the gain or loss need to be present?</span></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="sc">&gt;</span> proportion_threshold <span class="ot">&lt;-</span> <span class="fl">0.2</span></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a></span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a><span class="co"># Run command</span></span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a><span class="sc">&gt;</span> output <span class="ot">&lt;-</span> <span class="fu">MakeSummaryTable</span>(rcn.obj, </span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a>                             low_threshold, high_threshold, </span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a>                             prob_loss, prob_gain, </span>
<span id="cb6-22"><a href="#cb6-22" tabindex="-1"></a>                             proportion_threshold,</span>
<span id="cb6-23"><a href="#cb6-23" tabindex="-1"></a>                             <span class="at">find_peaks =</span> <span class="cn">TRUE</span>)</span>
<span id="cb6-24"><a href="#cb6-24" tabindex="-1"></a><span class="sc">&gt;</span> output<span class="sc">$</span>summary_table <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">5</span>)</span>
<span id="cb6-25"><a href="#cb6-25" tabindex="-1"></a><span class="co"># A tibble: 5 × 16</span></span>
<span id="cb6-26"><a href="#cb6-26" tabindex="-1"></a>  chromosome    start      end mean_copy_number gain_probability loss_probability gain_count</span>
<span id="cb6-27"><a href="#cb6-27" tabindex="-1"></a>  <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>         <span class="er">&lt;</span>int<span class="sc">&gt;</span>    <span class="er">&lt;</span>int<span class="sc">&gt;</span>            <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>            <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>            <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>      <span class="er">&lt;</span>dbl<span class="sc">&gt;</span></span>
<span id="cb6-28"><a href="#cb6-28" tabindex="-1"></a><span class="dv">1</span> <span class="dv">1</span>           <span class="dv">8910001</span>  <span class="dv">9060000</span>            <span class="fl">0.157</span>            <span class="fl">0.244</span>           <span class="fl">0.0049</span>          <span class="dv">5</span></span>
<span id="cb6-29"><a href="#cb6-29" tabindex="-1"></a><span class="dv">2</span> <span class="dv">1</span>           <span class="dv">9600001</span> <span class="dv">10230000</span>            <span class="fl">0.139</span>            <span class="fl">0.252</span>           <span class="fl">0.0038</span>          <span class="dv">6</span></span>
<span id="cb6-30"><a href="#cb6-30" tabindex="-1"></a><span class="dv">3</span> <span class="dv">1</span>          <span class="dv">10620001</span> <span class="dv">10740000</span>            <span class="fl">0.119</span>            <span class="fl">0.222</span>           <span class="fl">0.0062</span>          <span class="dv">5</span></span>
<span id="cb6-31"><a href="#cb6-31" tabindex="-1"></a><span class="dv">4</span> <span class="dv">1</span>          <span class="dv">15420001</span> <span class="dv">15630000</span>            <span class="fl">0.148</span>            <span class="fl">0.292</span>           <span class="fl">0.106</span>           <span class="dv">3</span></span>
<span id="cb6-32"><a href="#cb6-32" tabindex="-1"></a><span class="dv">5</span> <span class="dv">1</span>          <span class="dv">15960001</span> <span class="dv">16080000</span>            <span class="fl">0.155</span>            <span class="fl">0.302</span>           <span class="fl">0.106</span>           <span class="dv">3</span></span>
<span id="cb6-33"><a href="#cb6-33" tabindex="-1"></a><span class="co"># ℹ 9 more variables: loss_count &lt;dbl&gt;, gain_proportion &lt;dbl&gt;, loss_proportion &lt;dbl&gt;,</span></span>
<span id="cb6-34"><a href="#cb6-34" tabindex="-1"></a><span class="co">#   gain_samples &lt;chr&gt;, loss_samples &lt;chr&gt;, bin_count &lt;int&gt;, sum_of_bin_lengths &lt;dbl&gt;,</span></span>
<span id="cb6-35"><a href="#cb6-35" tabindex="-1"></a><span class="co">#   coordinates &lt;chr&gt;, features_in_region &lt;chr&gt;</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="v---scaling-to-the-absolute-copy-number-space">V - Scaling to the Absolute Copy-Number Space<a class="anchor" aria-label="anchor" href="#v---scaling-to-the-absolute-copy-number-space"></a>
</h2>
<div class="section level3">
<h3 id="v-1---performing-the-scaling">V.1 - Performing the scaling<a class="anchor" aria-label="anchor" href="#v-1---performing-the-scaling"></a>
</h3>
<p>The Relative CNs space can be useful for identifying trends of
relative gains and losses across samples in an experiment. However,
absolute copy-numbers (aCNs) can be even more useful as they indicate
ploidy and place real numbers on genomic features such as
high-amplification events. <strong>Utanos</strong> makes use of the <a href="https://github.com/crukci-bioinformatics/rascal" class="external-link">rascal</a>
package to scale from the relative to absolute space.<br>
Here, we will briefly detail the options and workflow, but for a more
thorough understanding we recommend the user read through the rascal
vignette.</p>
<p>There are three major options for scaling depending on the data
available:</p>
<ol style="list-style-type: decimal">
<li>
<strong><em>mad/rmsd</em></strong>: Without any other data/metadata,
just the copy-numbers, this is the best option. Different
ploidy/cellularity combinations are assumed, the relative CNs are scaled
to absolute, and then the difference of each value to the nearest
whole-number is calculated. The difference is quantified as either the
mean absolute deviation (mad) or root mean squared error (rmsq).
Finally, a grid search is performed to determine the local minima and
these are reported as the solution(s). If several ‘solutions’ are nearly
equivalent then rascal returns several, otherwise a single
cellularity-ploidy-mad/rmsq distance combo is returned. Many more
options for tuning are available in the package and will not be
discussed here.</li>
<li>
<strong><em>vafs</em></strong>: In the case of having variant allele
frequencies (vafs) for each sample, these can be used to identify the
nearest ploidy-cellularity combination (as described in 1). There are
two options under this umbrella. One will assume the maximum discovered
VAF for the sample is an appropriate representation for the tumour
fraction. Below is a equation describing the relationship used to select
the optimal solution from the grid search. For more details see the
vignette in the <a href="https://github.com/crukci-bioinformatics/rascal" class="external-link">rascal github
repo</a>.</li>
</ol>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>t</mi><mi>u</mi><mi>m</mi><mi>o</mi><mi>u</mi><mi>r</mi><mi>f</mi><mi>r</mi><mi>a</mi><mi>c</mi><mi>t</mi><mi>i</mi><mi>o</mi><msub><mi>n</mi><mi>i</mi></msub><mspace width="0.222em"></mspace><mo>=</mo><mspace width="0.222em"></mspace><mfrac><mrow><msub><mi>a</mi><mi>i</mi></msub><mi>c</mi></mrow><mrow><msub><mi>a</mi><mi>i</mi></msub><mi>c</mi><mo>+</mo><mn>2</mn><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>−</mo><mi>c</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></mfrac></mrow><annotation encoding="application/x-tex">tumour fraction_i \ = \ \frac{a_i c}{a_i c + 2(1-c)}</annotation></semantics></math></p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>c</mi><mo>=</mo><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi><mi>u</mi><mi>l</mi><mi>a</mi><mi>r</mi><mi>i</mi><mi>t</mi><mi>y</mi><mrow><mtext mathvariant="normal">, </mtext><mspace width="0.333em"></mspace></mrow><msub><mi>a</mi><mi>i</mi></msub><mo>=</mo><mtext mathvariant="normal">copy-number at locus i</mtext></mrow><annotation encoding="application/x-tex"> c = cellularity \text{, } a_i = \text{copy-number at locus i}</annotation></semantics></math></p>
<p>The other will use the vafs of only certain genes the user
identifies. For example if the user expects ‘PTEN’ vafs to be suggestive
of the tumour fraction they can choose to scale using just those
vafs.</p>
<ol start="3" style="list-style-type: decimal">
<li>
<strong><em>Known ploidies</em></strong>: In the case of knowing the
sample ploidies via some other method (ex. FISH), this option will
simply scale the relative copy-numbers for each sample to absolute given
those ploidies.</li>
</ol>
<p>The <strong>utanos</strong> function documentation contains specifics
on expected formatting for each of these options.</p>
<p><strong>Utanos</strong> splits aCN scaling into two parts. First, the
likeliest solutions (ploidy/cellularity/distance) are calculated using
rascal, and then second the values are scaled and a new CN-object is
made.</p>
<div class="section level4">
<h4 id="v-1-1---performing-the-scaling-part-1">V.1.1 - Performing the scaling (part 1)<a class="anchor" aria-label="anchor" href="#v-1-1---performing-the-scaling-part-1"></a>
</h4>
<hr>
<p>To find the optimal ploidy and cellularity, use the
<code><a href="../reference/FindRascalSolutions.html">FindRascalSolutions()</a></code> function.<br>
Only the first argument is mandatory, the rest are optional.</p>
<p>Arguments:</p>
<ol style="list-style-type: decimal">
<li>
<code>cnobj</code> - The relative copy-number object we’ve been
using throughout all these vignettes.<br>
</li>
<li>
<code>min_ploidy</code> &amp; <code>max_ploidy</code> - Determines
the range of ploidies to search from.<br>
</li>
<li>
<code>ploidy_step</code> - The stepwise increment of ploidies along
the range.<br>
</li>
<li>
<code>min_cellularity</code> &amp; <code>max_cellularity</code> -
Determines the range of cellularities to search from.<br>
</li>
<li>
<code>cellularity_step</code> - The stepwise increment of
cellularities along the range.<br>
</li>
<li>
<code>distance_function</code> - The distance function to use, “MAD”
(mean absolute difference) or “RMSD” (root mean square
difference).<br>
</li>
<li>
<code>distance_filter_scale_factor</code> - The distance threshold
above which solutions will be discarded as a multiple of the solution
with the smallest distance.<br>
</li>
<li>
<code>max_proportion_zero</code> - The maximum proportion of fitted
absolute copy number values in the zero copy number state.<br>
</li>
<li>
<code>in_proportion_close_to_whole_number</code> - The minimum
proportion of fitted absolute copy number values sufficiently close to a
whole number.<br>
</li>
<li>
<code>max_distance_from_whole_number</code> - The maximum distance
from a whole number that a fitted absolute copy number can be to be
considered sufficiently close.<br>
</li>
<li>
<code>solution_proximity_threshold</code> - How close two solutions
can be before one will be filtered; reduces the number of best fit
solutions where there are many minima in close proximity.<br>
</li>
<li>
<code>keep_all</code> - <code>TRUE</code> to return all solutions
but with additional <code>best_fit</code> column to indicate which are
the local minima that are acceptable solutions (may be useful to avoid
computing the distance grid twice).</li>
</ol>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="sc">&gt;</span> solutions <span class="ot">&lt;-</span> <span class="fu">FindRascalSolutions</span>(cnobj)</span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="v-1-2---performing-the-scaling-part-2">V.1.2 - Performing the scaling (part 2)<a class="anchor" aria-label="anchor" href="#v-1-2---performing-the-scaling-part-2"></a>
</h4>
<p>To scale the samples use the <code><a href="../reference/CalculateACNs.html">CalculateACNs()</a></code>
function.<br>
The first three arguments are mandatory, the rest are optional depending
on context.<br>
The first 4 arguments would benefit from some elaboration, so those will
be described, and then we provide some examples for using this
function.</p>
<p>Arguments:</p>
<ol style="list-style-type: decimal">
<li>
<code>cnobj</code> - The relative copy-number object we’ve been
using throughout all these vignettes.<br>
</li>
<li>
<code>acnmethod</code> - The method for scaling. Expects a string.
It can be one of:
<ul>
<li>“maxvaf”: Chooses the solution that aligns with the maximum detected
VAF. We assume that to be suggestive of the cellularity according to the
formula above</li>
<li>A character vector of the genes from which to pull VAFs, where genes
are assumed to be in order of decreasing precedence, ex. c(‘TP53’,
‘KRAS’, ‘PTEN’),</li>
<li>“mad”: Chooses the solution with the lowest MAD or RMSD score; in
the case of a tie goes with the lower ploidy solution.</li>
<li>A dataframe of known ploidies. Scales each sample to aCN space using
these plodies. Expects a column with <code>sample_id</code> - matching
those in the <code>rcn.obj</code>, and <code>ploidy</code> - a numeric
value.</li>
</ul>
</li>
<li>
<code>rascal_sols</code> - The third argument expects a TSV file or
data.frame of the calculated rascal solutions. See part 1 for details on
creation.</li>
</ol>
<p>Example solutions table:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="sc">&gt;</span> solutions <span class="ot">&lt;-</span> <span class="fu">FindRascalSolutions</span>(cnobj)</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">head</span>(solutions)</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="do">##      sample_id ploidy cellularity distance</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="do">## 1: CC-CHM-1341   3.34        0.85    0.166</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="do">## 2: CC-CHM-1341   1.34        0.46    0.166</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="do">## 3: CC-CHM-1341   3.33        0.90    0.166</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="do">## 4: CC-CHM-1341   2.34        0.60    0.166</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="do">## 5: CC-CHM-1341   2.35        0.52    0.174</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="do">## 6: CC-CHM-1341   4.34        1.00    0.176</span></span></code></pre></div>
<ol start="4" style="list-style-type: decimal">
<li>
<code>variants</code> - The fourth argument for CalculateACNs, is an
optional argument. In the event of having variants data this parameter
expects a table or file path to a comma-separated values (CSV) file
containing long-format data.<br>
Requirements for the variants table when using VAFs from targeted panel
sequencing or some other technology:
<ul>
<li>The variants must must be in a datatable/dataframe</li>
<li>The required columns are: <code>sample_id</code>,
<code>chromosome</code>, <code>start</code>, <code>end</code>,
<code>gene_name</code>, <code>ref</code>, <code>alt</code>,
<code>vaf</code> (with these exact names, but in any order)</li>
<li>Each row of said table must correspond to a unique variant</li>
<li>Each variant must have an associated variant allele frequency</li>
<li>Each row must also be associated with a specific sample,
<code>sample_id</code> must match</li>
</ul>
</li>
</ol>
<p>Example variants data table:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="sc">&gt;</span> variants_path <span class="ot">&lt;-</span> <span class="st">"~/some_path/variants.csv"</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="sc">&gt;</span> variants <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="at">file =</span> variants_path, <span class="at">sep =</span> <span class="st">','</span>)</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="sc">&gt;</span> variants <span class="ot">&lt;-</span> variants <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">gene_name =</span> gene_symbol)</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">head</span>(variants[, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="do">##      sample_id chromosome     start       end ref alt gene_name</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="do">## 1: CC-CHM-1341          3  30691871  30691872  GA   G    TGFBR2</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="do">## 2: CC-CHM-1341          3 142274739 142274740  AT   A       ATR</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="do">## 3: CC-CHM-1341          7 151945071 151945071   G  GT     KMT2C</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a><span class="do">## 4: CC-CHM-1341         10  90771767  90771767   G   A       FAS</span></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="do">## 5: CC-CHM-1341         16    343635    343635   C   T     AXIN1</span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a><span class="do">## 6: CC-CHM-1341         17   7577556   7577556   C   T      TP53</span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a><span class="do">##                       consequence estimated_impact  vafs</span></span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a><span class="do">## 1:             frameshift_variant             HIGH 0.201</span></span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a><span class="do">## 2:             frameshift_variant             HIGH 0.196</span></span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a><span class="do">## 3: stop_gained,frameshift_variant             HIGH 0.212</span></span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a><span class="do">## 4:               missense_variant         MODERATE 0.301</span></span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a><span class="do">## 5:               missense_variant         MODERATE 0.226</span></span>
<span id="cb9-18"><a href="#cb9-18" tabindex="-1"></a><span class="do">## 6:               missense_variant         MODERATE 0.720</span></span></code></pre></div>
<p>Note: this example table has more information than is necessary.
There are extraneous columns.</p>
<p><br></p>
<p>Examples:</p>
<p>A. Scaling using the solution with the best (lowest) distance:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="sc">&gt;</span> outputs <span class="ot">&lt;-</span> <span class="fu">CalculateACNs</span>(rcn.obj, </span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>                           <span class="at">acn.method =</span> <span class="st">'mad'</span>, </span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>                           <span class="at">rascal_sols =</span> solutions)</span></code></pre></div>
<p>B. Scaling using the ‘maxvaf’ option and providing some variant
allele frequencies</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="sc">&gt;</span> outputs <span class="ot">&lt;-</span> <span class="fu">CalculateACNs</span>(rcn.obj, </span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>                           <span class="at">acn.method =</span> <span class="st">'maxvaf'</span>, </span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>                           <span class="at">rascal_sols =</span> solutions, </span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>                           <span class="at">variants =</span> variants, </span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>                           <span class="at">return_sols =</span> <span class="cn">TRUE</span>, </span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>                           <span class="at">return_S4 =</span> <span class="cn">TRUE</span>)</span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>Warning message<span class="sc">:</span></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>In <span class="fu">ReplaceQDNAseqAssaySlots</span>(cnobj, output[[<span class="st">"acns"</span>]], output[[<span class="st">"acns_segs"</span>]]) <span class="sc">:</span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a>  Not all samples had ACN solutions.</span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a>    These samples are excluded from the returned acn object.</span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a>    The probability of loss<span class="sc">/</span>gain values <span class="cf">in</span> the returned acn object were calculated using the relative CNs object.</span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">names</span>(outputs)</span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"rascal_solutions"</span> <span class="st">"acn.obj"</span></span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a><span class="sc">&gt;</span> outputs<span class="sc">$</span>rascal_solutions <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">5</span>)</span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a>    sample_id ploidy cellularity tumour_fraction   vaf</span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a><span class="dv">1</span> CC<span class="sc">-</span>CHM<span class="dv">-1341</span>   <span class="fl">1.34</span>        <span class="fl">0.46</span>           <span class="fl">0.365</span>  <span class="fl">0.72</span></span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a><span class="dv">2</span> CC<span class="sc">-</span>CHM<span class="dv">-1347</span>   <span class="fl">3.46</span>        <span class="fl">0.32</span>           <span class="fl">0.592</span> <span class="fl">0.874</span></span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a><span class="dv">3</span> CC<span class="sc">-</span>CHM<span class="dv">-1355</span>      <span class="dv">2</span>        <span class="fl">0.83</span>           <span class="fl">0.828</span> <span class="fl">0.395</span></span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a><span class="dv">4</span> CC<span class="sc">-</span>CHM<span class="dv">-1361</span>      <span class="dv">2</span>        <span class="fl">0.72</span>           <span class="fl">0.719</span> <span class="fl">0.313</span></span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a><span class="dv">5</span> CC<span class="sc">-</span>HAM<span class="dv">-0369</span>   <span class="fl">1.94</span>           <span class="dv">1</span>               <span class="dv">1</span> <span class="fl">0.872</span></span></code></pre></div>
<p>C. Scaling by prioritizing variants from certain genes.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="sc">&gt;</span> outputs <span class="ot">&lt;-</span> <span class="fu">CalculateACNs</span>(rcn.obj, </span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>                           <span class="at">acnmethod =</span> <span class="fu">c</span>(<span class="st">"TP53"</span>, <span class="st">"KRAS"</span>, <span class="st">"PTEN"</span>), </span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>                           <span class="at">rascal_sols =</span> solutions, </span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>                           <span class="at">variants =</span> variants, </span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a>                           <span class="at">return_sols =</span> <span class="cn">TRUE</span>, </span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a>                           <span class="at">return_S4 =</span> <span class="cn">TRUE</span>)</span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a>Warning message<span class="sc">:</span></span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a>In <span class="fu">ReplaceQDNAseqAssaySlots</span>(cnobj, output[[<span class="st">"acns"</span>]], output[[<span class="st">"acns_segs"</span>]]) <span class="sc">:</span></span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a>  Not all samples had ACN solutions.</span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a>    These samples are excluded from the returned acn object.</span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a>    The probability of loss<span class="sc">/</span>gain values <span class="cf">in</span> the returned acn object were calculated using the relative CNs object.</span></code></pre></div>
<p>Note: The warning message for C provides a useful message to users in
case an ACN solution for some samples was not found by rascal.</p>
</div>
</div>
<div class="section level3">
<h3 id="v-2---plotting">V.2 - Plotting<a class="anchor" aria-label="anchor" href="#v-2---plotting"></a>
</h3>
<div class="section level4">
<h4 id="v-2-1---copy-number-segment-plots">V.2.1 - Copy-Number Segment Plots<a class="anchor" aria-label="anchor" href="#v-2-1---copy-number-segment-plots"></a>
</h4>
<p>These scaled copy-numbers, absolute copy-numbers (aCNs), can be
visualized using the same plotting function mentioned in vignette 1.
Let’s briefly compare the aCNs generated using the maxvaf solution
compared to the ‘mad’ solution.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="sc">&gt;</span> maxvaf_outputs <span class="ot">&lt;-</span> <span class="fu">CalculateACNs</span>(rcn.obj, </span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>                                  <span class="at">acn.method =</span> <span class="st">'maxvaf'</span>, </span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>                                  <span class="at">rascal_sols =</span> solutions, </span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>                                  <span class="at">variants =</span> variants, </span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>                                  <span class="at">return_S4 =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a><span class="sc">&gt;</span> mad_outputs <span class="ot">&lt;-</span> <span class="fu">CalculateACNs</span>(rcn.obj, </span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a>                               <span class="at">acnmethod =</span> <span class="st">'mad'</span>, </span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a>                               <span class="at">rascal_sols =</span> solutions, </span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a>                               <span class="at">return_S4 =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">CNSegmentsPlot</span>(maxvaf_outputs<span class="sc">$</span>acn.obj, </span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a>                 <span class="at">sample =</span> <span class="st">"CC-CHM-1341"</span>, </span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a>                 <span class="at">min_copy_number =</span> <span class="dv">0</span>, </span>
<span id="cb13-13"><a href="#cb13-13" tabindex="-1"></a>                 <span class="at">max_copy_number =</span> <span class="fl">12.5</span>, </span>
<span id="cb13-14"><a href="#cb13-14" tabindex="-1"></a>                 <span class="at">copy_number_breaks =</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>))</span></code></pre></div>
<p><img src="images/cndiversity_vignette_03.png" width="100%"></p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">CNSegmentsPlot</span>(mad_outputs<span class="sc">$</span>acn.obj, </span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>                 <span class="at">sample =</span> <span class="st">"CC-CHM-1341"</span>, </span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>                 <span class="at">min_copy_number =</span> <span class="dv">0</span>, </span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a>                 <span class="at">max_copy_number =</span> <span class="fl">12.5</span>, </span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a>                 <span class="at">copy_number_breaks =</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>))</span></code></pre></div>
<p><img src="images/cndiversity_vignette_04.png" width="100%"></p>
<p>The scaling using the ‘mad’ set the sample ploidy to around 4 whereas
the ‘maxvaf’ option instead preferred a ploidy of 2.</p>
</div>
<div class="section level4">
<h4 id="v-2-2---copy-number-diversity-plots">V.2.2 - Copy-Number Diversity Plots<a class="anchor" aria-label="anchor" href="#v-2-2---copy-number-diversity-plots"></a>
</h4>
<p>Finally, similarly to section III.2, samples can be visualized all
together as a heatmap. Using this strategy, sample-to-sample
similarities and differences are easily differentiable. For datasets
with larger numbers of samples, it is recommended to order the samples
in some informative manner. This is done by default in the
<code>ACNDiversityPlot</code> function. The ordering is found by
performing hierarchical clustering on the passed in copy-number
segments.</p>
<p>By design, the <code>ACNDiversityPlot</code> function expects data in
long-format. This is to encourage users to think about the bin-size they
would like to plot. For example, when plotting a genome 3 billion base
pairs long bins 30kb across will barely be visible to the eye. A bin
this size on a 7-inch plot would make up roughly 7e-05 of the width. So
even if there is a one-bin amplification, it will not be visible on the
heatmap. However, it will matter a lot the the plotting software.
Drawing rectangles for this many minuscule bins potentially across many
samples will be take a very long time!<br>
TL;DR there is no real point in plotting tiny bins for this
plot-type.<br>
Instead, re-shape the data to a comfortable bin-size for plotting to
improve speed. We have found 1e06 to 2e06 to be reasonably quick for up
to several hundred samples.<br>
If the user is interested in finding tiny amplification events,
filtering and examining the tabular outputs would be a much better
option (ex. <code>outputs$acn_segment_tables</code>).</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="sc">&gt;</span> outputs <span class="ot">&lt;-</span> <span class="fu">CalculateACNs</span>(rcn.obj, </span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>                           <span class="at">acnmethod =</span> <span class="st">'mad'</span>, </span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>                           <span class="at">rascal_sols =</span> solutions)</span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="sc">&gt;</span> long_segs <span class="ot">&lt;-</span> <span class="fu">SegmentsToCopyNumber</span>(outputs<span class="sc">$</span>acn_segment_tables, </span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a>                                    <span class="dv">1000000</span>, </span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a>                                    <span class="at">genome =</span> <span class="st">'hg19'</span>, </span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a>                                    <span class="at">Xincluded =</span> <span class="cn">TRUE</span>)</span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">colnames</span>(long_segs) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"chromosome"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"state"</span>, <span class="st">"sample_id"</span>)</span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">ACNDiversityPlot</span>(<span class="at">long_segments =</span> long_segs)</span></code></pre></div>
<p><img src="images/cndiversity_vignette_05.png" width="100%"></p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Maxwell Douglas, Branden Lynch, Carlos Vasquez, Jacky Yiu, Ding Ma, Nirupama Tamvada, Sameer Shankar.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
