<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>3. Investigating and Creating Copy-Number Signatures • utanos</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="3. Investigating and Creating Copy-Number Signatures">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">utanos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/utanos.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/A-samplequality-vignette.html">1. Sample Filtering and Quality Evaluation</a></li>
    <li><a class="dropdown-item" href="../articles/B-cndiversityandscaling-vignette.html">2. Investigating Copy-Number Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/C-cnsignatures-vignette.html">3. Investigating and Creating Copy-Number Signatures</a></li>
    <li><a class="dropdown-item" href="../articles/D-shallowhrd-vignette.html">4. Predicting Homologous Recombination Deficiency (HRD)</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>3. Investigating and Creating Copy-Number Signatures</h1>
            
      

      <div class="d-none name"><code>C-cnsignatures-vignette.Rmd</code></div>
    </div>

    
    
<hr>
<div class="section level3">
<h3 id="pre-amble">Pre-amble<a class="anchor" aria-label="anchor" href="#pre-amble"></a>
</h3>
<hr>
<p>This vignette demonstrates how to use copy-number (CN) abnormalities
data, which could represent the imprints of distinct mutational
processes, to either build new CN-Signatures or determine which existing
signatures are present in a dataset. This workflow is based on one
described in the methods section of a <a href="https://www.nature.com/articles/s41588-018-0179-8" class="external-link">paper
published</a> in 2018 <span class="citation">(Macintyre et al.
2018)</span>.</p>
<p>Most of the published literature thus far has focused on
CN-Signatures from absolute copy-number data. That is to say data where
the sample ploidy is known. However, CN-Signatures can be extracted from
both relative and absolute CN data, and so both cases will be covered
here.</p>
<p>First, this vignette will cover finding signature exposures in a new
dataset using existing signatures (section III). Second, the creation of
new signatures will be covered (section IV). In either case, to begin, a
set of samples with segmented copy-number profiles are needed. These
profiles should be either formatted as a list of named dataframes or
exist within a <code>QDNAseq</code> object.</p>
<hr>
</div>
<div class="section level3">
<h3 id="i---input-data">I - Input data<a class="anchor" aria-label="anchor" href="#i---input-data"></a>
</h3>
<hr>
<p>In earlier vignettes it sufficed to use just a few samples to
demonstrate the usage of utanos functions. However, in order to generate
new cn-signatures, considerably more data is needed. Therefore, to
demonstrate both signature exposure calling and the creation of new
signatures, the Britroc HGSOC (High-Grade Serous Ovarian Carcinoma)
cohort will be used <span class="citation">(Macintyre et al.
2018)</span>. The segmented copy-number data can be found <a href="https://bitbucket.org/britroc/cnsignatures/src/master/manuscript_Rmarkdown/data/" class="external-link">here</a>
in the paper’s bitbucket repository.</p>
<p>So step one, download said data/clone the repo, and read in the
absolute copy-numbers:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(utanos)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(dplyr)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="sc">&gt;</span> acn.obj <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"~/repos/cnsignatures/manuscript_Rmarkdown/data/britroc_absolute_copynumber.rds"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="sc">&gt;</span> acn.obj</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="fu">QDNAseqCopyNumbers</span> (storageMode<span class="sc">:</span> lockedEnvironment)</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>assayData<span class="sc">:</span> <span class="dv">103199</span> features, <span class="dv">280</span> samples </span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>  element names<span class="sc">:</span> copynumber, segmented </span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>protocolData<span class="sc">:</span> none</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>phenoData</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>  sampleNames<span class="sc">:</span> IM_100 IM_101 ... JBLAB<span class="dv">-4282</span> (<span class="dv">280</span> total)</span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>  varLabels<span class="sc">:</span> name total.reads ... <span class="fu">mapTP53cn</span> (<span class="dv">10</span> total)</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a>  varMetadata<span class="sc">:</span> labelDescription</span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a>featureData</span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a>  featureNames<span class="sc">:</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">1-30000</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30001-60000</span> ... Y<span class="sc">:</span><span class="dv">59370001-59373566</span> (<span class="dv">103199</span> total)</span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a>  fvarLabels<span class="sc">:</span> chromosome start ... <span class="fu">use</span> (<span class="dv">9</span> total)</span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a>  fvarMetadata<span class="sc">:</span> labelDescription</span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a>experimentData<span class="sc">:</span> use <span class="st">'experimentData(object)'</span></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a>Annotation<span class="sc">:</span>  </span></code></pre></div>
<p>For this vignette we will use just the higher-quality samples (2 and
3 star in the context of that paper). A list is available in
utanosmodellingdata.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">britroc_samples_subset</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/data.table/man/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"~/repos/utanosmodellingdata/sample_sets/brit117_samples.csv"</span><span class="op">)</span></span>
<span><span class="va">acn.obj</span> <span class="op">&lt;-</span> <span class="va">acn.obj</span><span class="op">[</span>,<span class="va">britroc_samples_subset</span><span class="op">$</span><span class="va">x</span><span class="op">]</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="ii---extract-cn-features">II - Extract CN-features<a class="anchor" aria-label="anchor" href="#ii---extract-cn-features"></a>
</h3>
<hr>
<p>Step 2: To model copy-number aberrations in the genome, find trends
and commonalities (what we will call signatures henceforth), specific
features within the dataset are pulled out and modeled separately. We
can extract copy number features using either the
<code><a href="../reference/ExtractCopyNumberFeatures.html">ExtractCopyNumberFeatures()</a></code> or
<code><a href="../reference/ExtractRelativeCopyNumberFeatures.html">ExtractRelativeCopyNumberFeatures()</a></code> functions.</p>
<div class="section level4">
<h4 id="ii-1---relative-copy-numbers">II.1 - Relative Copy-Numbers<a class="anchor" aria-label="anchor" href="#ii-1---relative-copy-numbers"></a>
</h4>
<p>Section coming soon!</p>
</div>
<div class="section level4">
<h4 id="ii-2---relative-copy-numbers">II.2 - Relative Copy-Numbers<a class="anchor" aria-label="anchor" href="#ii-2---relative-copy-numbers"></a>
</h4>
<p>Standard features:</p>
<ol style="list-style-type: decimal">
<li>Absolute copy-numbers at each genomic bin (copynumbers)<br>
</li>
<li>The magnitude of each copy number change (changepoints)<br>
</li>
<li>The genomic length (in bp) of each segment (segsizes)<br>
</li>
<li>The number of breakpoints per chromosome arm (brchrmarm)<br>
</li>
<li>The number of breakpoints per 10 megabases (bp10MB)<br>
</li>
<li>The genomic length (in bp) of oscillating copy number regions
(osCNs)</li>
</ol>
<p>Optional features:</p>
<ol start="7" style="list-style-type: decimal">
<li>Minimum number of chromosomes (a count) needed to account for 50% of
CN changes in a sample (nc50)<br>
</li>
<li>Distance in base pairs of each breakpoint to the centromere
(cdist)</li>
</ol>
<p>Both of the two CN-Signature sets included in utanosmodellingdata
(from p53abn endometrial and HGSOC) use the standard features. For a few
hundred samples, this function should run within a few minutes.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cn_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractCopyNumberFeatures.html">ExtractCopyNumberFeatures</a></span><span class="op">(</span><span class="va">acn.obj</span>, </span>
<span>                                         genome <span class="op">=</span> <span class="st">"hg19"</span>, </span>
<span>                                         extra_features <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="iii---determine-cn-signature-exposures">III - Determine CN-signature exposures<a class="anchor" aria-label="anchor" href="#iii---determine-cn-signature-exposures"></a>
</h3>
<hr>
<p>The HGSOC CN-Signatures included in the utanosmodellingdata
repository were created using 91 samples (the highest quality ones). To
find exposures for the remaining samples in the group, subset the
original object by those samples not used for signature creation, and
calculate exposures.</p>
<p>The <code><a href="../reference/CallSignatureExposures.html">CallSignatureExposures()</a></code> function in utanos runs
several computational steps in the appropriate order for convenience. A
user can run each of these separately should they want. For details see
the docs for
<code><a href="../reference/ExtractCopyNumberFeatures.html">ExtractCopyNumberFeatures()</a></code>/<code><a href="../reference/ExtractRelativeCopyNumberFeatures.html">ExtractRelativeCopyNumberFeatures()</a></code>,
<code><a href="../reference/GenerateSampleByComponentMatrix.html">GenerateSampleByComponentMatrix()</a></code>, and
<code><a href="../reference/QuantifySignatures.html">QuantifySignatures()</a></code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hgsoc_sigs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"~/repos/utanosmodellingdata/signatures/30kb_ovarian/component_by_signature_britroc_aCNs.rds"</span><span class="op">)</span></span>
<span><span class="va">acn.obj.subset</span> <span class="op">&lt;-</span> <span class="va">acn.obj</span><span class="op">[</span>,<span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">acn.obj</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">hgsoc_sigs</span><span class="op">@</span><span class="va">consensus</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">sigexposures</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CallSignatureExposures.html">CallSignatureExposures</a></span><span class="op">(</span>copy_numbers_input <span class="op">=</span> <span class="va">acn.obj.subset</span>, </span>
<span>                                       component_models <span class="op">=</span> <span class="st">"~/repos/utanosmodellingdata/component_models/30kb_ovarian/component_models_britroc_aCNs.rds"</span>, </span>
<span>                                       signatures <span class="op">=</span> <span class="st">"~/repos/utanosmodellingdata/signatures/30kb_ovarian/component_by_signature_britroc_aCNs.rds"</span>,</span>
<span>                                       refgenome <span class="op">=</span> <span class="st">"hg19"</span>, </span>
<span>                                       relativeCN_data <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </span></code></pre></div>
<p>The exposures can be visualized with the
<code><a href="../reference/SignatureExposuresPlot.html">SignatureExposuresPlot()</a></code> function like so:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ggp1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SignatureExposuresPlot.html">SignatureExposuresPlot</a></span><span class="op">(</span><span class="va">sigexposures</span><span class="op">)</span></span>
<span><span class="va">ggp1</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<p><img src="images%2Fsignatures_vignette_01.png" width="100%"></p>
<p>Alongside the plot, the <code><a href="../reference/SignatureExposuresPlot.html">SignatureExposuresPlot()</a></code>
function returns the sample ordering in the plot. Given that these
signatures are from the Macintyre et al. 2018 publication, let’s
re-order to match the signature order from the paper for easy
comparison:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reord_britroc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">6</span>,<span class="fl">5</span>,<span class="fl">4</span>,<span class="fl">7</span>,<span class="fl">3</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ggp1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SignatureExposuresPlot.html">SignatureExposuresPlot</a></span><span class="op">(</span><span class="va">sigexposures</span><span class="op">[</span><span class="va">reord_britroc</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="va">ggp1</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<p><img src="images%2Fsignatures_vignette_02.png" width="100%"></p>
<p>Interpreting this plot, it is clear the sample subset has higher
exposures for signatures 1 and 4. Referencing <span class="citation">(Macintyre et al. 2018)</span>, these signatures
correspond to poor overall survival + breakage-fusion bridge events and
whole-genome duplication + PI3K/AKT signalling inactivation
respectively.</p>
<p><br></p>
<hr>
</div>
<div class="section level3">
<h3 id="iv---create-new-cn-signatures">IV - Create New CN-Signatures<a class="anchor" aria-label="anchor" href="#iv---create-new-cn-signatures"></a>
</h3>
<hr>
<p>Creating copy-number signatures anew involves a few more steps than
directly determining exposures from existing signatures. This process
can also be iterative where given some metrics, earlier steps are re-run
with different parameters/inputs. This vignette will be using the
Britroc cohort samples to demonstrate signature creation but will not be
seeking to re-create the signatures found in that 2018 paper exactly.
For an exact replication, the user to will need to set all parameters
exactly as seen in the paper repository and run a few steps without the
utanos wrappers.<br>
To reiterate, the modelling and signatures seen below will not match
<span class="citation">(Macintyre et al. 2018)</span>. This is
expected.</p>
<div class="section level4">
<h4 id="iv-1---modelling">IV.1 - Modelling<a class="anchor" aria-label="anchor" href="#iv-1---modelling"></a>
</h4>
<p>First, the CN features discussed in section II are modeled using
mixtures of gaussians or poissons with the help of the flexmix R-package
<span class="citation">(Grün and Leisch 2008)</span>. Utanos provides a
wrapper called <code><a href="../reference/FitMixtureModels.html">FitMixtureModels()</a></code>. Then, given these
mixture models for each CN feature, all the samples used for modelling
are compared to each component of each model and a sum-of-posterior
probabilities matrix is calculated. The
<code><a href="../reference/GenerateSampleByComponentMatrix.html">GenerateSampleByComponentMatrix()</a></code> utanos function
calculates these posteriors. In more detail: For each copy-number event
for a feature in each sample, compute the posterior probability of
belonging to a component for said feature. For each sample, sum these
posterior event vectors to a sum-of-posterior probabilities vector.
Combine the sum-of-posterior vectors into a patient-by-component
sum-of-posterior probabilities matrix.</p>
<p>Finally, given the patient-by-component sum-of-posterior
probabilities matrix, matrix decomposition is run many times to decide
on an optimal rank for the two output matrices (sample-by-signature
matrix and a signature-by-component matrix). The optimal rank is the
optimal number of signatures. The NMF R-package is leveraged for matrix
decomposition using the <code><a href="../reference/ChooseNumberSignatures.html">ChooseNumberSignatures()</a></code> function
<span class="citation">(Gaujoux and Seoighe 2010)</span>.</p>
<p>Succinctly, the modelling stage involves…</p>
<ol style="list-style-type: decimal">
<li>Fitting mixture components.<br>
</li>
<li>Calculating a patient-by-component sum-of-posterior probabilities
matrix.</li>
<li>Searching for the optimal number of signatures (interval of 3-12 is
used by default), running NMF many times (ex. 1000x) with different
random seeds for each signature number.</li>
</ol>
<p><strong>Note on 2.</strong> The returned sample_by_component matrix
may have the components in a different order than the components in the
component_models object. The components are not in increasing order in
the component_models object.</p>
<p><strong>Note on 3.</strong> Estimating the optimal number of
signatures is quite computationally heavy. This function takes the
longest to run of any in the utanos package. It might make sense to run
this function and then get a snack!</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Model using just the highest quality samples</span></span>
<span><span class="va">acn.obj.high</span> <span class="op">&lt;-</span> <span class="va">acn.obj</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">acn.obj</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">hgsoc_sigs</span><span class="op">@</span><span class="va">consensus</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">cnfeatures_high</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractCopyNumberFeatures.html">ExtractCopyNumberFeatures</a></span><span class="op">(</span><span class="va">acn.obj.high</span>, </span>
<span>                                             genome <span class="op">=</span> <span class="st">"hg19"</span>, </span>
<span>                                             extra_features <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">component_models</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FitMixtureModels.html">FitMixtureModels</a></span><span class="op">(</span><span class="va">cnfeatures_high</span><span class="op">)</span></span>
<span><span class="va">sample_by_component</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GenerateSampleByComponentMatrix.html">GenerateSampleByComponentMatrix</a></span><span class="op">(</span><span class="va">CN_features</span>, <span class="va">component_models</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The next function contains a method that MUST have NMF loaded, this is a known issue with the NMF package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://renozao.github.io/NMF/" class="external-link">NMF</a></span><span class="op">)</span></span>
<span><span class="va">rank_estimate_outputs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ChooseNumberSignatures.html">ChooseNumberSignatures</a></span><span class="op">(</span><span class="va">sample_by_component</span><span class="op">)</span></span>
<span><span class="fu">NMF</span><span class="fu">::</span><span class="fu">plot</span><span class="op">(</span><span class="va">rank_estimate_outputs</span><span class="op">$</span><span class="va">estrank</span>, <span class="va">rank_estimate_outputs</span><span class="op">$</span><span class="va">randomized_estrank</span>,</span>
<span>          what <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cophenetic"</span>, <span class="st">"dispersion"</span>, <span class="st">"sparseness"</span>, <span class="st">"silhouette"</span><span class="op">)</span>,</span>
<span>          xname <span class="op">=</span> <span class="st">"Observed"</span>, yname <span class="op">=</span> <span class="st">"Randomised"</span>, main <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span></code></pre></div>
<p><img src="images%2Fsignatures_vignette_03.png" width="80%"></p>
<p>This dot and line plot contains several metrics from the NMF runs
over possible ranks that aid in the selection of an optimal number of
signatures.</p>
<p>Brief heuristics for what to look for in the values of each
metric:</p>
<table class="table">
<colgroup>
<col width="27%">
<col width="44%">
<col width="27%">
</colgroup>
<thead><tr class="header">
<th align="center">Metric</th>
<th align="center">Description</th>
<th align="center">Interpretation</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">cophenetic</td>
<td align="center">after the trend begins increasing choose the max
value before it decreases again.</td>
<td align="center">higher =&gt; better</td>
</tr>
<tr class="even">
<td align="center">dispersion</td>
<td align="center">measures the reproducibility</td>
<td align="center">higher =&gt; better</td>
</tr>
<tr class="odd">
<td align="center">silhouette</td>
<td align="center">measures consistency between clusters</td>
<td align="center">higher =&gt; better</td>
</tr>
<tr class="even">
<td align="center">sparseness</td>
<td align="center">choose the maximum sparsity which can be achieved
without that for the corresponding randomly permuted matrix exceeding
it</td>
<td align="center">higher =&gt; better</td>
</tr>
</tbody>
</table>
<p>For detailed discussions about each of these metrics consult the <a href="https://renozao.github.io/NMF/devel/index.html" class="external-link">NMF</a> package
documentation or other sources online.</p>
<p>The utanos package defaults for the
<code><a href="../reference/ChooseNumberSignatures.html">ChooseNumberSignatures()</a></code> function include running NMF 100
times for each rank. This was to keep the default run-time manageable.
Package users however are encouraged to experiment with more iterations
and other run parameter values. The <span class="citation">(Macintyre et
al. 2018)</span> paper for instance used 1000 iterations in estimating
the rank (this and a few other parameters are the reason this plot
differs from the one in the paper bitbucket repository).</p>
<p><br></p>
</div>
<div class="section level4">
<h4 id="iv-2---evaluation">IV.2 - Evaluation<a class="anchor" aria-label="anchor" href="#iv-2---evaluation"></a>
</h4>
<p>At this point it is a good idea to stop and examine the modelling
done thus far.<br>
Some points to consider:</p>
<ul>
<li>If the metrics provided above for the input matrices do not appear
to be outperforming the randomized matrices, you may wish to set more
restrictive demands on what samples can be used as input data. The
ability to detect copy number signatures depends heavily on whether each
pattern is sufficiently represented in the data, BUT, this signal can be
obfuscated with lower quality samples that introduce noise. Finding a
balance of quantity vs. quality is important. The largest number of
high-quality samples possible used for signature creation (but no more)
is ideal.</li>
<li>By default utanos suggests using 6 CN features for modelling,
however it may be desirable to use more/others. There are two more that
can easily be used by setting the <code>extra_features</code> parameter
of <code><a href="../reference/ExtractCopyNumberFeatures.html">ExtractCopyNumberFeatures()</a></code> to TRUE.</li>
<li>Examine the sample-by-component matrix as a heatmap to see if there
are major contributions from many different samples to each component.
If the sample-by-component matrix does not indicate higher values for
the posteriors for at least several components per sample it could be
the case that the modelling is not appropriately capturing the
heterogeneity in the data.</li>
<li>Examine the consensus matrices (the average connectivity matrix
across NMF runs) in heatmap form alongside the silhouette track (or
others) to see how samples cluster together. Do clear clusters
emerge?</li>
</ul>
<p>Visualize the sample-by-component matrix:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">NMF</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/NMF/man/aheatmap.html" class="external-link">aheatmap</a></span><span class="op">(</span><span class="va">sample_by_component</span>, fontsize <span class="op">=</span> <span class="fl">7</span>, </span>
<span>              Rowv<span class="op">=</span><span class="cn">FALSE</span>, Colv<span class="op">=</span><span class="cn">FALSE</span>, cexRow <span class="op">=</span> <span class="fl">0</span>, legend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>              breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">199</span>,<span class="fl">2</span><span class="op">)</span>,<span class="fl">500</span><span class="op">)</span>, </span>
<span>              main<span class="op">=</span><span class="st">"Component x Sample matrix"</span><span class="op">)</span></span></code></pre></div>
<p><img src="images%2Fsignatures_vignette_04.png" width="80%"></p>
<p><br></p>
<p>And the consensus matrices across 4 ranks (5,6,7, and 8):</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">NMF</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/NMF/man/heatmaps.html" class="external-link">consensusmap</a></span><span class="op">(</span><span class="va">rank_estimate</span><span class="op">$</span><span class="va">fit</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span><span class="op">]</span>, main<span class="op">=</span><span class="st">'Consensus matrix - Cluster stability'</span>, treeheight <span class="op">=</span> <span class="fl">0</span>, labCol<span class="op">=</span><span class="cn">NA</span>, labRow<span class="op">=</span><span class="cn">NA</span>, tracks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"silhouette:"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="images%2Fsignatures_vignette_05.png" width="100%"></p>
<p>Note that the map corresponding to ranks 6 and 7 look pretty good,
the same samples tend to cluster together across runs. Samples tend to
be either dark red or dark blue rather than a lighter shade somewhere in
between. For more details on connectivity/consensus matrices see the
docs for <code><a href="https://rdrr.io/pkg/NMF/man/connectivity.html" class="external-link">NMF::connectivity()</a></code>.</p>
<p>If content with the modelling and a clear factorization rank (number
of signatures) emerges, move on to the next section and create a
signature set.</p>
<p><br></p>
</div>
<div class="section level4">
<h4 id="iv-3---create-cn-signatures">IV.3 - Create CN-Signatures<a class="anchor" aria-label="anchor" href="#iv-3---create-cn-signatures"></a>
</h4>
<p>The <code><a href="../reference/GenerateSignatures.html">GenerateSignatures()</a></code> function is again running NMF
many times (but not over multiple ranks) and therefore takes the
second-longest to run of any function in the utanos package. It might
make sense to start this function and then go get a snack!</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">signatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GenerateSignatures.html">GenerateSignatures</a></span><span class="op">(</span><span class="va">sample_by_component</span>, <span class="fl">7</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Usually worth stopping and saving the objects just created since re-running can take a while. </span></span>
<span><span class="co"># saveRDS(signatures, file = paste0("path_to_signatures/signatures_object.rds"))</span></span>
<span><span class="co"># saveRDS(component_models, file = paste0("path_to_signatures/component_models_object.rds"))</span></span></code></pre></div>
<p>The <code><a href="../reference/GenerateSignatures.html">GenerateSignatures()</a></code> function is very simple,
wrapping just a single NMF call. It provides some default parameter
values (ex. sets seed = 77777, and the nmf algorithm used to “brunet”)
but should the user which to test many different options use the
<code><a href="https://rdrr.io/pkg/NMF/man/nmf.html" class="external-link">NMF::nmf()</a></code> function directly.</p>
<p><br></p>
<p>As a final step to creating new signatures, visualize them using the
NMF package’s various heatmap options:</p>
<ul>
<li>signature-by-component matrix as a heatmap (what we call a
‘signature set’ or ‘signatures’)</li>
<li>sample-by-sample consensus matrix as a heatmap (same as above, just
for one rank)</li>
<li>sample-by-signature matrix as a heatmap (contribution of each sample
to each signature)</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/NMF/man/heatmaps.html" class="external-link">basismap</a></span><span class="op">(</span><span class="va">signatures</span>, Rowv <span class="op">=</span> <span class="cn">NA</span>, main <span class="op">=</span> <span class="st">"Signature x Component matrix"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/NMF/man/heatmaps.html" class="external-link">consensusmap</a></span><span class="op">(</span><span class="va">signatures</span>, main<span class="op">=</span><span class="st">'Consensus matrix - Cluster stability'</span>, </span>
<span>             treeheight <span class="op">=</span> <span class="fl">0</span>, labRow <span class="op">=</span> <span class="cn">NA</span>, labCol <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/NMF/man/heatmaps.html" class="external-link">coefmap</a></span><span class="op">(</span><span class="va">signatures</span>, Colv<span class="op">=</span><span class="st">"consensus"</span>, tracks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"basis:"</span><span class="op">)</span>, </span>
<span>        main<span class="op">=</span><span class="st">"Sample x Signature matrix"</span>, treeheight <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="images%2Fsignatures_vignette_06.png" width="60%"></p>
<p><img src="images%2Fsignatures_vignette_07.png" width="60%"></p>
<p><img src="images%2Fsignatures_vignette_08.png" width="60%"></p>
</div>
<div class="section level4">
<h4 id="section-note">Section Note<a class="anchor" aria-label="anchor" href="#section-note"></a>
</h4>
<p>For the modelling section utanos provides a number of wrapping
functions for flexmix and NMF for convenience. However, it’s not
unrealistic that the user may want more flexibility than these wrappers
allow. In that case it is quite straightforward to grab the utanos
source code from github out of these functions and run each step
individually with more control.</p>
<p><br></p>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="v---compare-signature-sets">V - Compare signature sets<a class="anchor" aria-label="anchor" href="#v---compare-signature-sets"></a>
</h3>
<hr>
<p>Utanos includes 3 methods by which to compare between different
signature sets.</p>
<ol style="list-style-type: decimal">
<li>Using the wasserstein distance (as a heatmap) to compare mixture
components for two CN-features.</li>
<li>Compare mixture components for two CN-features using shaded line
plots</li>
<li>Alluvial flow or Sankey plots to visulalize the difference in how
samples will cluster given their exposures to two signature sets. By
default the samples are assigned the signature for which they have the
maximum exposure.</li>
</ol>
<p>For this section, comparisons will be made between the signatures
just created above, the signatures from the 2018 MacIntyre paper
(HGSOC), and the 2024 Jamieson paper (p53abn EC).</p>
<p>cm_ex = example component models made in this vignette<br>
sigs_ex = example signatures made in this vignette<br>
cm_hgsoc = Component models from the 2018 Macintyre paper<br>
sigs_hgsoc = Signatures from the 2018 Macintyre paper<br>
cm_p53ec = Components models created using samples from the 2024
Jamieson paper<br>
sigs_p53ec = Signatures from the 2024 Jamieson paper</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cm_ex</span> <span class="op">&lt;-</span> <span class="va">component_models</span></span>
<span><span class="va">sigs_ex</span> <span class="op">&lt;-</span> <span class="va">signatures</span></span>
<span><span class="va">cm_hgsoc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"~/repos/utanosmodellingdata/component_models/30kb_ovarian/component_models_britroc_aCNs.rds"</span><span class="op">)</span></span>
<span><span class="va">sigs_hgsoc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"~/repos/utanosmodellingdata/signatures/30kb_ovarian/component_by_signature_britroc_aCNs.rds"</span><span class="op">)</span></span>
<span><span class="va">cm_p53ec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"~/repos/utanosmodellingdata/component_models/30kb_endometrial/component_models_vancouver_aCNs.rds"</span><span class="op">)</span></span>
<span><span class="va">sigs_p53ec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"~/repos/utanosmodellingdata/signatures/30kb_endometrial/component_by_signature_britmodelsvansigs5_aCNs.rds"</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="v-1---earth-movers-distance">V.1 - Earth-movers distance<a class="anchor" aria-label="anchor" href="#v-1---earth-movers-distance"></a>
</h4>
<p>One-to-one sequential comparisons of each component composing two
different signatures is difficult. There is no guarantee that a
signature set run with one set of parameters or with one input dataset
will have the same number of components as with another another set of
parameters or input dataset. Therefore, comparisons that are
intrinsically many-to-many are a better option.</p>
<p>Utanos provides a function, <code><a href="../reference/WassDistancePlot.html">WassDistancePlot()</a></code>, that
calculates the 2-D Wasserstein (earth-movers distance) to transform each
component in a signature for a given feature to each other component for
the same feature in another signature set. A higher distance means two
components are less similar and a lower distance means the opposite. The
<code><a href="../reference/WassDistancePlot.html">WassDistancePlot()</a></code> function then orders the components in
increasing order and colours high-similarity components lightly and less
similar components a darker hue.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span> <span class="co"># For easy plotting of multiple figures next to one another</span></span>
<span><span class="va">cm_ex</span> <span class="op">&lt;-</span> <span class="va">component_models</span></span>
<span><span class="va">sigs_ex</span> <span class="op">&lt;-</span> <span class="va">signatures</span></span>
<span><span class="va">ggp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/WassDistancePlot.html">WassDistancePlot</a></span><span class="op">(</span>cm_a <span class="op">=</span> <span class="va">cm_ex</span>, </span>
<span>                         cm_b <span class="op">=</span> <span class="va">cm_hgsoc</span>, </span>
<span>                         component <span class="op">=</span> <span class="st">'osCN'</span><span class="op">)</span></span>
<span><span class="va">ggp3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/WassDistancePlot.html">WassDistancePlot</a></span><span class="op">(</span>cm_a <span class="op">=</span> <span class="va">cm_ex</span>, </span>
<span>                         cm_b <span class="op">=</span> <span class="va">cm_hgsoc</span>, </span>
<span>                         component <span class="op">=</span> <span class="st">'copynumber'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ggp2</span><span class="op">$</span><span class="va">plot</span> <span class="op">+</span> <span class="va">ggp3</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<p><img src="images%2Fsignatures_vignette_09.png" width="90%"></p>
<p>The earth-movers distance heatmap shows that the two lowest
components from the example signature’s oscillating CN-feature are both
highly similar to the first component in the published signatures. The
CN-feature ‘copynumber’ heatmap shows a typicaly trend of the lower
components being much more similar to one another and then the
distributions flatten and extend as they rise in value. Note that
components A - c.6 and B - c.8 are less similar than we might expect.
This is likely due to there being just 6 components in in the A set
(example set), and so the value is pulled down.</p>
</div>
<div class="section level4">
<h4 id="v-2---shaded-line-plots-of-mixture-components">V.2 - Shaded line plots of mixture components<a class="anchor" aria-label="anchor" href="#v-2---shaded-line-plots-of-mixture-components"></a>
</h4>
<p>Individual mixture components for a signature can be visualized using
shaded line plots. In the case of the “segment size” copy-number feature
this will be a series of Gaussians. The <code><a href="../reference/MixtureModelPlots.html">MixtureModelPlots()</a></code>
function from utanos can return plots for any or all of the components
that compose each signature. Additionally, the specific components that
are most prevalent for the signature indicated are shaded.</p>
<p>Some examples.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Reorder the HGSOC signatures the same way as is done in the paper (keep things easy to compare)</span></span>
<span><span class="va">sigs_hgsoc_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NMF/man/basis-coef-methods.html" class="external-link">basis</a></span><span class="op">(</span><span class="va">sigs_hgsoc</span><span class="op">)</span></span>
<span><span class="va">reord_britroc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">6</span>,<span class="fl">5</span>,<span class="fl">4</span>,<span class="fl">7</span>,<span class="fl">3</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sigs_hgsoc_df</span> <span class="op">&lt;-</span> <span class="va">sigs_hgsoc_df</span><span class="op">[</span>,<span class="va">reord_britroc</span><span class="op">]</span></span>
<span></span>
<span><span class="va">mixture_plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MixtureModelPlots.html">MixtureModelPlots</a></span><span class="op">(</span><span class="va">sigs_hgsoc_df</span>, </span>
<span>                                   <span class="va">cm_hgsoc</span>, </span>
<span>                                   sig_of_interest <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span>
<span><span class="va">mixture_plots</span><span class="op">$</span><span class="va">copynumber</span></span></code></pre></div>
<p><img src="images%2Fsignatures_vignette_10.png" width="60%"></p>
<p>As briefly mentioned in section V.2, the CN-feature ‘copynumber’
shows a typical trend of lower components being much more similar to one
another and then the distributions flatten and extend as they rise in
value reflecting an increase in variance. This is easy to see in the
shaded line plots.</p>
</div>
<div class="section level4">
<h4 id="v-3---alluvial-flow-plots-of-samples-across-signatures">V.3 - Alluvial flow plots of samples across signatures<a class="anchor" aria-label="anchor" href="#v-3---alluvial-flow-plots-of-samples-across-signatures"></a>
</h4>
<p>To make an Alluvial-flow/Sankey plot 2 comparisons are needed.</p>
<p>Signature exposures for each sample to each set of signatures:<br>
Britroc HGSOC samples -&gt; Britroc HGSOC Sigs<br>
Britroc HGSOC samples -&gt; Vancouver p53abn Endometrial Sigs</p>
<p>This is result in two matrices of signature exposures per sample.</p>
<p>The utanos function <code><a href="../reference/SEAlluvialPlot.html">SEAlluvialPlot()</a></code> takes two matrices
of signature exposures and creates an Alluvial plot using the
easyalluvial R-package. Of course, the same samples must be in both
matrices. If not already installed, install the easyalluvial
package.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/erblast/easyalluvial/" class="external-link">easyalluvial</a></span><span class="op">)</span></span>
<span><span class="co"># Pull exposures for the 91 highest quality samples out of the signatures object itself (the coefficients matrix)</span></span>
<span><span class="va">expo_mat_hq</span> <span class="op">&lt;-</span> <span class="fu">NMF</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/NMF/man/basis-coef-methods.html" class="external-link">scoef</a></span><span class="op">(</span><span class="va">sigs_hgsoc</span><span class="op">)</span></span>
<span><span class="va">expo_mat_hq</span> <span class="op">&lt;-</span> <span class="va">expo_mat_hq</span><span class="op">[</span><span class="va">reord_britroc</span>,<span class="op">]</span></span>
<span><span class="va">expo_mat_hq</span> <span class="op">&lt;-</span> <span class="fu">NormaliseMatrix</span><span class="op">(</span><span class="va">expo_mat_hq</span><span class="op">)</span></span>
<span><span class="co"># Exposures were calculated for the 117-91 samples back in Section 3. </span></span>
<span><span class="va">expo_mat_mq</span> <span class="op">&lt;-</span> <span class="va">sigexposures</span><span class="op">[</span><span class="va">reord_britroc</span>,<span class="op">]</span></span>
<span><span class="co"># Combine</span></span>
<span><span class="va">expo_hgsoc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">expo_mat_hq</span>, <span class="va">expo_mat_mq</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Find exposures of HGSOC samples to p53abn EC signatures</span></span>
<span><span class="va">expo_ec</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CallSignatureExposures.html">CallSignatureExposures</a></span><span class="op">(</span><span class="va">acn.obj</span>,</span>
<span>                                  component_models <span class="op">=</span> <span class="va">cm_hgsoc</span>, </span>
<span>                                  signatures <span class="op">=</span> <span class="va">sigs_p53ec</span>, </span>
<span>                                  refgenome <span class="op">=</span> <span class="st">'hg19'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Pass both matrices of signature exposures to the SEAlluvialPlot utanos function</span></span>
<span><span class="va">alluvial_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SEAlluvialPlot.html">SEAlluvialPlot</a></span><span class="op">(</span><span class="va">expo_hgsoc</span>, <span class="va">expo_ec</span><span class="op">)</span></span>
<span><span class="va">alluvial_output</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<p><img src="images%2Fsignatures_vignette_11.png" width="60%"></p>
<p>Notably, the alluvial plot highlights the likely similarity of
signature 1 in HGSOC set and signature 2 in the p53abn EC set. Many
samples with a maximum exposure to one, also have a max. exposure to the
other. The same goes for A4 and B3.</p>
<p>Each of the past three subsections in combination with tests of
significance can help the user in comparing and contrasting copy-number
signature sets.</p>
<p><br></p>
</div>
<div class="section level4 unnumbered">
<h4 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h4>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-nmf2010" class="csl-entry">
Gaujoux, Renaud, and Cathal Seoighe. 2010. <span>“A Flexible r Package
for Nonnegative Matrix Factorization.”</span> <em>BMC
Bioinformatics</em> 11 (1): 367. <a href="https://doi.org/10.1186/1471-2105-11-367" class="external-link">https://doi.org/10.1186/1471-2105-11-367</a>.
</div>
<div id="ref-flexmix2008" class="csl-entry">
Grün, Bettina, and Friedrich Leisch. 2008. <span>“<span>FlexMix</span>
Version 2: Finite Mixtures with Concomitant Variables and Varying and
Constant Parameters.”</span> <em>Journal of Statistical Software</em> 28
(4): 1–35. <a href="https://doi.org/10.18637/jss.v028.i04" class="external-link">https://doi.org/10.18637/jss.v028.i04</a>.
</div>
<div id="ref-macintyre2018copy" class="csl-entry">
Macintyre, Geoff, Teodora E Goranova, Dilrini De Silva, Darren Ennis,
Anna M Piskorz, Matthew Eldridge, Daoud Sie, et al. 2018. <span>“Copy
Number Signatures and Mutational Processes in Ovarian Carcinoma.”</span>
<em>Nature Genetics</em> 50 (9): 1262–70.
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Maxwell Douglas, Branden Lynch, Carlos Vasquez, Jacky Yiu, Matthew Eldridge, Geoff Macintyre, Ding Ma, Nirupama Tamvada, Sameer Shankar.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
