<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>4. Predicting Homologous Recombination Deficiency (HRD) • utanos</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="4. Predicting Homologous Recombination Deficiency (HRD)">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">utanos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/utanos.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/A-samplequality-vignette.html">1. Sample Filtering and Quality Evaluation</a></li>
    <li><a class="dropdown-item" href="../articles/B-cndiversityandscaling-vignette.html">2. Investigating Copy-Number Diversity</a></li>
    <li><a class="dropdown-item" href="../articles/C-cnsignatures-vignette.html">3. Investigating and Creating Copy-Number Signatures</a></li>
    <li><a class="dropdown-item" href="../articles/D-shallowhrd-vignette.html">4. Predicting Homologous Recombination Deficiency (HRD)</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>4. Predicting Homologous Recombination Deficiency (HRD)</h1>
            
      

      <div class="d-none name"><code>D-shallowhrd-vignette.Rmd</code></div>
    </div>

    
    
<hr>
<div class="section level3">
<h3 id="i---introduction-background">I - Introduction &amp; background<a class="anchor" aria-label="anchor" href="#i---introduction-background"></a>
</h3>
<hr>
<p>This vignette discusses the prediction of homologous recombination
deficiency (HRD) from relative copy-number (rCN) data derived from
shallow whole-genome sequencing (sWGS). Detecting HRD is of interest to
researchers and clinicians, particularly in ovarian and breast cancer,
as treatment of HRD cases with poly (ADP-ribose) polymerase inhibitors
(PARPis) is associated with improved progression-free survival <span class="citation">(Davies et al. 2017; Callens et al. 2023)</span>. As
HRD tumors are partially, or entirely, unable to repair double-strand
breaks (DSBs) via homologous recombination, they rely on alternative
pathways for DSB repair. This reliance is reflected in characteristic
‘genomic scars’ in HRD tumors, and various academic and commercial
assays have attempted to use these scars, often in conjunction with
mutations in key HR-related genes such as BRCA1/2, to identify HRD in
patient tumors <span class="citation">(Davies et al. 2017; Callens et
al. 2023)</span>.</p>
<p>Many existing methods used to identify HRD status such as HRDetect
<span class="citation">(Davies et al. 2017)</span> or the commercial
Myriad MyChoice assay require variant calling, necessitating deep
whole-genome sequencing or targeted sequencing in conjunction with
another method of CN assessment, such as sWGS or a SNP array. Given the
significantly reduced cost of sWGS, it is desirable to predict HRD
solely from CN data furnished by sWGS. In 2020 Eeckhoutte et al. devised
shallowHRD <span class="citation">(Eeckhoutte et al. 2020)</span>, a
tool for the prediction of HRD status from CN data derived from sWGS,
which has since been validated in at least one follow-up study <span class="citation">(Guo et al. 2023)</span>. An updated version of
shallowHRD has since been published <span class="citation">(Callens et
al. 2023)</span>, but the code is not publicly available.</p>
<p>In the interest of streamlining HRD prediction, we implemented
version one of shallowHRD <span class="citation">(Eeckhoutte et al.
2020)</span> in Utanos, with some changes to ease the ingestion of
QDNAseq objects and cohort-wide HRD prediction; the core algorithm
remains the same. In brief, large-genomic alterations (LGAs), defined as
adjacent segments in the same arm that are greater than 10 Mb in size
and separated by fewer than 3 Mb, are one example of a genomic scar that
is indicative of HRD <span class="citation">(Callens et al.
2023)</span>. ShallowHRD seeks to identify LGAs in an iterative fashion,
composed of two passes through the segmented CN data. Because of noise
that is inherent to the sampling process, which varies as a function of
read depth, among other factors like GC-content, mappability, and
artifacts derived from fixation, segmentation is necessarily imperfect.
ShallowHRD therefore seeks to merge adjacent segments which differ in CN
by less than a threshold in order to better identify LGAs. In the first
pass, kernel density estimation (KDE) is used to find a local minimum in
the segment-wise CN distance – that is, the CN distance between adjacent
segments. Adjacent segments with CN distance less that the identified
threshold are merged and their segmented CN values are recalculated
accordingly. After merging segments in the first pass, a second estimate
of the optimal merging threshold is determined in the same fashion, this
time using the updated segmented CN values produced by the first pass.
This provides a refined estimate of the merging threshold, which is then
used to merge the original segments, yielding the final set of merged
segments. Finally, the LGAs in the sample are counted, and the HRD
status is determined; samples with 20 or more LGAs are considered HRD,
with the remainder being non-HRD.</p>
<hr>
</div>
<div class="section level3">
<h3 id="ii---input-data">II - Input data<a class="anchor" aria-label="anchor" href="#ii---input-data"></a>
</h3>
<hr>
<p>As an example dataset we will again use the samples found in the
<em>utanosmodellingdata</em> repository found <a href="https://github.com/Huntsmanlab/utanosmodellingdata" class="external-link">here</a>. If
not already done, clone that repo to somewhere convenient such as a
common ‘repos’ folder on your machine and read in the example data. It
is human endometrial carcinoma sWGS data aligned to hg19. The associated
publication <span class="citation">(Jamieson et al. 2024)</span> was
published in CCR and can be found <a href="https://doi.org/10.1158/1078-0432.CCR-23-3689" class="external-link">here</a>.</p>
<p>Download said data/clone the repo, and read in the relative
copy-numbers:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(utanos)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(QDNAseq)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(ggplot2)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="sc">&gt;</span> rcn.obj <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"~/repos/utanosmodellingdata/sample_copynumber_data/sample_rcn_data.rds"</span>)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="sc">&gt;</span> rcn.obj</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">QDNAseqCopyNumbers</span> (storageMode<span class="sc">:</span> lockedEnvironment)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>assayData<span class="sc">:</span> <span class="dv">103199</span> features, <span class="dv">10</span> samples </span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>  element names<span class="sc">:</span> copynumber, segmented </span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>protocolData<span class="sc">:</span> none</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>phenoData</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>  sampleNames<span class="sc">:</span> CC<span class="sc">-</span>CHM<span class="dv">-1341</span> CC<span class="sc">-</span>CHM<span class="dv">-1347</span> ... CC<span class="sc">-</span>HAM<span class="dv">-0385</span> (<span class="dv">10</span> total)</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a>  varLabels<span class="sc">:</span> name total.reads ... <span class="fu">loess.family</span> (<span class="dv">6</span> total)</span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>  varMetadata<span class="sc">:</span> labelDescription</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>featureData</span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a>  featureNames<span class="sc">:</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">1-30000</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30001-60000</span> ... Y<span class="sc">:</span><span class="dv">59370001-59373566</span> (<span class="dv">103199</span> total)</span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a>  fvarLabels<span class="sc">:</span> chromosome start ... <span class="fu">comCNV.mask</span> (<span class="dv">11</span> total)</span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a>  fvarMetadata<span class="sc">:</span> labelDescription</span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a>experimentData<span class="sc">:</span> use <span class="st">'experimentData(object)'</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a>Annotation<span class="sc">:</span> </span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="iii---extract-data-and-run-shallowhrd">III - Extract data and run shallowHRD<a class="anchor" aria-label="anchor" href="#iii---extract-data-and-run-shallowhrd"></a>
</h3>
<hr>
<p>We can run shallowHRD directly on the QDNAseq object we just read in,
using the <code><a href="../reference/RunShallowHRDFromQDNA.html">RunShallowHRDFromQDNA()</a></code> function, or on a table
of segments, using the <code><a href="../reference/RunShallowHRD.html">RunShallowHRD()</a></code> function. Since
estimating the threshold used to merge segments requires 100,000
iterations by default, running a single sample can take a few minutes.
In the interest of time, we will extract a single sample from the
QDNAseq object and demonstrate running shallowHRD on a table of
segments. To run it for all the samples in the object, we could just
call <code><a href="../reference/RunShallowHRDFromQDNA.html">RunShallowHRDFromQDNA()</a></code> directly on the QDNAseq
object, with most arguments remaining the same as those we’ll provide to
<code><a href="../reference/RunShallowHRD.html">RunShallowHRD()</a></code>.</p>
<p>Let’s extract the bin-wise and segment-wise rCN data form the QDNAseq
object using the <code><a href="../reference/ExportBinsQDNAObj.html">ExportBinsQDNAObj()</a></code>, and join them into a
single table. This process is taken care of for us by
<code><a href="../reference/RunShallowHRDFromQDNA.html">RunShallowHRDFromQDNA()</a></code> if running shallowHRD for all
samples.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="sc">&gt;</span> bin_df <span class="ot">&lt;-</span> <span class="fu">ExportBinsQDNAObj</span>(<span class="at">object =</span> rcn.obj, <span class="at">type =</span> <span class="st">"copynumber"</span>, <span class="at">filter =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>     tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">!</span><span class="fu">c</span>(<span class="st">"feature"</span>, <span class="st">"chromosome"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>), <span class="at">names_to =</span> <span class="st">"sample"</span>, <span class="at">values_to =</span> <span class="st">"ratio"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>     dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">!</span>feature)</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="sc">&gt;</span> seg_df <span class="ot">&lt;-</span> <span class="fu">ExportBinsQDNAObj</span>(<span class="at">object =</span> rcn.obj, <span class="at">type =</span> <span class="st">"segments"</span>, <span class="at">filter =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>     tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">!</span><span class="fu">c</span>(<span class="st">"feature"</span>, <span class="st">"chromosome"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>), <span class="at">names_to =</span> <span class="st">"sample"</span>, <span class="at">values_to =</span> <span class="st">"ratio_median"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>     dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">!</span>feature)</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="sc">&gt;</span> df_group <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">inner_join</span>(<span class="at">x =</span> bin_df, <span class="at">y =</span> seg_df, <span class="at">by =</span> dplyr<span class="sc">::</span><span class="fu">join_by</span>(sample, chromosome, start, end)) <span class="sc">%&gt;%</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>     dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">!</span>end) <span class="sc">%&gt;%</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>     dplyr<span class="sc">::</span><span class="fu">group_by</span>(sample)</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="sc">&gt;</span> df_list <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">group_split</span>(df_group, <span class="at">.keep =</span> <span class="cn">FALSE</span>)</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">names</span>(df_list) <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">group_keys</span>(df_group)<span class="sc">$</span>sample</span></code></pre></div>
<p>Below we can see the resulting dataframe for a single sample. Note
that <code><a href="../reference/ExportBinsQDNAObj.html">ExportBinsQDNAObj()</a></code> does not log transform the rCN
data, so the values are centered about 1, rather than 0. The ‘ratio’
column indicates the bin-wise rCN, and the ‘ratio_median’ column
indicates the segment-wise rCN. If you wish to pass your own raw tables
to <code><a href="../reference/RunShallowHRD.html">RunShallowHRD()</a></code>, from one of the many tools for
generating rCN profiles, you should format the data in the same fashion,
with the same column names.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">head</span>(df_list<span class="sc">$</span><span class="st">`</span><span class="at">CC-CHM-1341</span><span class="st">`</span>)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co"># A tibble: 6 × 4</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>  chromosome  start ratio ratio_median</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>  <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>       <span class="er">&lt;</span>int<span class="sc">&gt;</span> <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>        <span class="er">&lt;</span>dbl<span class="sc">&gt;</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="dv">1</span> <span class="dv">1</span>          <span class="dv">840001</span> <span class="fl">1.11</span>         <span class="fl">0.941</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="dv">2</span> <span class="dv">1</span>          <span class="dv">870001</span> <span class="fl">1.05</span>         <span class="fl">0.941</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="dv">3</span> <span class="dv">1</span>          <span class="dv">900001</span> <span class="fl">1.02</span>         <span class="fl">0.941</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="dv">4</span> <span class="dv">1</span>          <span class="dv">930001</span> <span class="fl">0.893</span>        <span class="fl">0.941</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="dv">5</span> <span class="dv">1</span>          <span class="dv">960001</span> <span class="fl">1.17</span>         <span class="fl">0.941</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="dv">6</span> <span class="dv">1</span>          <span class="dv">990001</span> <span class="fl">0.988</span>        <span class="fl">0.941</span></span></code></pre></div>
<p>shallowHRD expects rCN data to be log-transformed, so when we call
<code><a href="../reference/RunShallowHRD.html">RunShallowHRD()</a></code>, we need to indicate that the data should
be log-transformed, by setting <code>log_transform = TRUE</code>. Since
we know this sample is from a female, we will include the X-chromosome:
<code>include_chr_X = TRUE</code>. Because the sampling process during
threshold estimation relies on pseudo-random number generation, we have
the option of specifying a seed, which will ensure repeated runs of the
same sample will produce an identical result. We can do so with the
<code>seed</code> parameter, but a value is specified by default, and it
is not necessary to change it. We will leave the
<code>num_simulations</code> to the default value of 100,000. We don’t
recommend reducing this value, as it leads to instability in the
estimation of the segment merging threshold in our experience. By
setting <code>shrd_save_path = './'</code>, we will save a table
summarizing the large segments detected, and the count of large segments
for a range of sizes, to the current working directory. Lastly, we
specify the sample name, which will be used to label our files and
plots, <code>sample = 'CC-CHM-1341'</code>, and we indicate that we wish
to generate a plot illustrating the segment merging process,
<code>plot = TRUE</code>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="sc">&gt;</span> shrd_result <span class="ot">&lt;-</span> <span class="fu">RunShallowHRD</span>(<span class="at">raw_ratios_file =</span> <span class="fu">as.data.frame</span>(df_list<span class="sc">$</span><span class="st">`</span><span class="at">CC-CHM-1341</span><span class="st">`</span>), </span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>               <span class="at">log_transform =</span> <span class="cn">TRUE</span>,</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>               <span class="at">include_chr_X =</span> <span class="cn">TRUE</span>,</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>               <span class="at">shrd_save_path =</span> <span class="st">'./'</span>,</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>               <span class="at">sample =</span> <span class="st">'CC-CHM-1341'</span>,</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>               <span class="at">plot =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>When you run the code in the block above, you’ll observe that every
thousandth iteration of the sampling process for threshold determination
is printed out (omitted here).</p>
<hr>
</div>
<div class="section level3">
<h3 id="iv---assess-the-results">IV - Assess the results<a class="anchor" aria-label="anchor" href="#iv---assess-the-results"></a>
</h3>
<hr>
<p>Let’s take a look at the results returned in
<code>shrd_result</code>.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="sc">&gt;</span> shrd_result</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="sc">$</span>hrd_status</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>[<span class="dv">1</span>] <span class="cn">FALSE</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="sc">$</span>n_lga</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>  Size_LGA Number_LGA</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="dv">1</span>        <span class="dv">3</span>         <span class="dv">27</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="dv">2</span>        <span class="dv">4</span>         <span class="dv">23</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="dv">3</span>        <span class="dv">5</span>         <span class="dv">23</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="dv">4</span>        <span class="dv">6</span>         <span class="dv">20</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="dv">5</span>        <span class="dv">7</span>         <span class="dv">15</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="dv">6</span>        <span class="dv">8</span>         <span class="dv">14</span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="dv">7</span>        <span class="dv">9</span>         <span class="dv">14</span></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a><span class="dv">8</span>       <span class="dv">10</span>         <span class="dv">11</span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a><span class="dv">9</span>       <span class="dv">11</span>         <span class="dv">11</span></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="sc">$</span>lga</span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a>   index chr chr_arm     start       end ratio_median     size level WC</span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a><span class="dv">1</span>     <span class="dv">34</span>   <span class="dv">1</span>     <span class="fl">1.0</span>    <span class="dv">840001</span>  <span class="dv">32100000</span> <span class="sc">-</span><span class="fl">0.087733372</span> <span class="dv">31260000</span>     <span class="dv">1</span>  <span class="dv">0</span></span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a><span class="dv">2</span>      <span class="dv">2</span>   <span class="dv">1</span>     <span class="fl">1.0</span>  <span class="dv">32100001</span>  <span class="dv">44520000</span>  <span class="fl">0.932817103</span> <span class="dv">12420000</span>     <span class="dv">2</span>  <span class="dv">1</span></span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a><span class="dv">3</span>      <span class="dv">3</span>   <span class="dv">1</span>     <span class="fl">1.0</span>  <span class="dv">44520001</span> <span class="dv">121350000</span>  <span class="fl">0.039840265</span> <span class="dv">76830000</span>     <span class="dv">3</span>  <span class="dv">1</span></span>
<span id="cb6-22"><a href="#cb6-22" tabindex="-1"></a><span class="dv">4</span>      <span class="dv">4</span>   <span class="dv">1</span>     <span class="fl">1.5</span> <span class="dv">143760001</span> <span class="dv">171090000</span>  <span class="fl">0.531069493</span> <span class="dv">27330000</span>     <span class="dv">4</span>  <span class="dv">0</span></span>
<span id="cb6-23"><a href="#cb6-23" tabindex="-1"></a><span class="dv">5</span>      <span class="dv">5</span>   <span class="dv">1</span>     <span class="fl">1.5</span> <span class="dv">171090001</span> <span class="dv">249210000</span>  <span class="fl">1.075190314</span> <span class="dv">78120000</span>     <span class="dv">5</span>  <span class="dv">1</span></span>
<span id="cb6-24"><a href="#cb6-24" tabindex="-1"></a><span class="dv">22</span>    <span class="dv">22</span>   <span class="dv">7</span>     <span class="fl">7.5</span>  <span class="dv">62460001</span>  <span class="dv">79560000</span> <span class="sc">-</span><span class="fl">0.024736678</span> <span class="dv">17100000</span>    <span class="dv">22</span>  <span class="dv">0</span></span>
<span id="cb6-25"><a href="#cb6-25" tabindex="-1"></a><span class="dv">23</span>    <span class="dv">23</span>   <span class="dv">7</span>     <span class="fl">7.5</span>  <span class="dv">79560001</span> <span class="dv">159120000</span>  <span class="fl">0.358396262</span> <span class="dv">79560000</span>    <span class="dv">23</span>  <span class="dv">1</span></span>
<span id="cb6-26"><a href="#cb6-26" tabindex="-1"></a><span class="dv">25</span>    <span class="dv">25</span>   <span class="dv">8</span>     <span class="fl">8.5</span>  <span class="dv">47460001</span>  <span class="dv">67080000</span>  <span class="fl">0.002882509</span> <span class="dv">19620000</span>    <span class="dv">25</span>  <span class="dv">0</span></span>
<span id="cb6-27"><a href="#cb6-27" tabindex="-1"></a><span class="dv">26</span>    <span class="dv">26</span>   <span class="dv">8</span>     <span class="fl">8.5</span>  <span class="dv">67080001</span> <span class="dv">146280000</span>  <span class="fl">0.611644543</span> <span class="dv">79200000</span>    <span class="dv">26</span>  <span class="dv">1</span></span>
<span id="cb6-28"><a href="#cb6-28" tabindex="-1"></a><span class="dv">31</span>    <span class="dv">31</span>   <span class="dv">9</span>     <span class="fl">9.5</span>  <span class="dv">94230001</span> <span class="dv">119400000</span> <span class="sc">-</span><span class="fl">0.128156351</span> <span class="dv">25170000</span>    <span class="dv">31</span>  <span class="dv">0</span></span>
<span id="cb6-29"><a href="#cb6-29" tabindex="-1"></a><span class="dv">32</span>    <span class="dv">32</span>   <span class="dv">9</span>     <span class="fl">9.5</span> <span class="dv">119400001</span> <span class="dv">135480000</span>  <span class="fl">0.377956527</span> <span class="dv">16080000</span>    <span class="dv">32</span>  <span class="dv">1</span></span>
<span id="cb6-30"><a href="#cb6-30" tabindex="-1"></a><span class="dv">38</span>    <span class="dv">38</span>  <span class="dv">11</span>    <span class="fl">11.5</span>  <span class="dv">55050001</span>  <span class="dv">68220000</span> <span class="sc">-</span><span class="fl">0.015957574</span> <span class="dv">13170000</span>    <span class="dv">38</span>  <span class="dv">0</span></span>
<span id="cb6-31"><a href="#cb6-31" tabindex="-1"></a><span class="dv">39</span>    <span class="dv">39</span>  <span class="dv">11</span>    <span class="fl">11.5</span>  <span class="dv">68220001</span>  <span class="dv">79920000</span>  <span class="fl">0.641546029</span> <span class="dv">11700000</span>    <span class="dv">39</span>  <span class="dv">1</span></span>
<span id="cb6-32"><a href="#cb6-32" tabindex="-1"></a><span class="dv">40</span>    <span class="dv">40</span>  <span class="dv">11</span>    <span class="fl">11.5</span>  <span class="dv">79920001</span>  <span class="dv">91470000</span> <span class="sc">-</span><span class="fl">0.018878010</span> <span class="dv">11550000</span>    <span class="dv">40</span>  <span class="dv">1</span></span>
<span id="cb6-33"><a href="#cb6-33" tabindex="-1"></a><span class="dv">44</span>    <span class="dv">44</span>  <span class="dv">12</span>    <span class="fl">12.5</span>  <span class="dv">38460001</span>  <span class="dv">63780000</span>  <span class="fl">0.462575888</span> <span class="dv">25320000</span>    <span class="dv">44</span>  <span class="dv">0</span></span>
<span id="cb6-34"><a href="#cb6-34" tabindex="-1"></a><span class="dv">45</span>    <span class="dv">45</span>  <span class="dv">12</span>    <span class="fl">12.5</span>  <span class="dv">63780001</span>  <span class="dv">74400000</span>  <span class="fl">0.829443681</span> <span class="dv">10620000</span>    <span class="dv">45</span>  <span class="dv">1</span></span>
<span id="cb6-35"><a href="#cb6-35" tabindex="-1"></a><span class="dv">46</span>    <span class="dv">46</span>  <span class="dv">12</span>    <span class="fl">12.5</span>  <span class="dv">74400001</span> <span class="dv">133830000</span>  <span class="fl">0.055195654</span> <span class="dv">59430000</span>    <span class="dv">46</span>  <span class="dv">1</span></span>
<span id="cb6-36"><a href="#cb6-36" tabindex="-1"></a><span class="dv">56</span>    <span class="dv">56</span>  <span class="dv">18</span>    <span class="fl">18.5</span>  <span class="dv">18540001</span>  <span class="dv">41190000</span>  <span class="fl">0.437227739</span> <span class="dv">22650000</span>    <span class="dv">56</span>  <span class="dv">0</span></span>
<span id="cb6-37"><a href="#cb6-37" tabindex="-1"></a><span class="dv">57</span>    <span class="dv">57</span>  <span class="dv">18</span>    <span class="fl">18.5</span>  <span class="dv">41190001</span>  <span class="dv">78000000</span> <span class="sc">-</span><span class="fl">0.150400989</span> <span class="dv">36810000</span>    <span class="dv">57</span>  <span class="dv">1</span></span>
<span id="cb6-38"><a href="#cb6-38" tabindex="-1"></a></span>
<span id="cb6-39"><a href="#cb6-39" tabindex="-1"></a><span class="sc">$</span>plot</span></code></pre></div>
<p>We can see that a list is returned, stored in
<code>shrd_result</code>. The first element, <code>hrd_status</code>, is
a boolean that indicates whether this sample is predicted to be HRD or
not; in this case we see that this sample predicted to be non-HRD. Next
we see <code>n_lga</code>, a table that indicates the number of genomic
alterations of a given size (in Mb) or greater. For example, we can see
that after merging segments, this sample had 27 segments of size 3 Mb or
greater, which includes 11 segments of 10 Mb or greater. Since a sample
must have 20 or more segments of 10 Mb or greater with altered rCN for a
positive prediction of HRD, we can see why this sample is non-HRD. The
<code>lga</code> element is a table of all the segments greater than 10
Mb in size, including segments that are not considered LGAs under the
precise criteria for LGAs, which is why we see more segments in this
table than in the corresponding entry in the <code>n_lga</code> table.
Recall that to be considered an LGA, a segment must be greater than 10
Mb, preceded by another segment greater than 10 Mb in the same
chromosome arm, and the distance between those segments must be less
than 3 Mb. The <code>WC</code> column marks LGA segments with a 1, and
non-LGA segments with a 0. If you add up the number of 1s in the
<code>WC</code> column, you will see that they sum to 11, as indicated
in row 8 of the <code>n_lga</code> table. Finally, the <code>plot</code>
element provides a plot that illustrates the segment merging
process:</p>
<p><img src="images%2Fshallowhrd_vignette_1.png" width="100%"></p>
<p>In this plot we can see our usual rCN profile, log-transformed and
centered about 0. The thick green bars indicate the original
segmentation. To visualize the process of segment merging over two
passes, the medium thickness aquamarine bars illustrate the segments
after merging with the first threshold, and the thinnest bright green
bars show the final segmentation, after merging using the threshold
estimated during the second pass. We can see that many segments are
merged in the first pass, with only a few undergoing additional merging
or un-merging during the second pass, such as those near the middle of
chromosome 7.</p>
<p>Let’s take a quick look at a sample that is predicted to be
borderline for HRD (15-19 LGAs). We can run shallowHRD for it in the
same fashion as before.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="sc">&gt;</span> shrd_result_borderline <span class="ot">&lt;-</span> <span class="fu">RunShallowHRD</span>(<span class="at">raw_ratios_file =</span> <span class="fu">as.data.frame</span>(df_list<span class="sc">$</span><span class="st">`</span><span class="at">CC-HAM-0385</span><span class="st">`</span>), </span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>               <span class="at">log_transform =</span> <span class="cn">TRUE</span>,</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>               <span class="at">include_chr_X =</span> <span class="cn">TRUE</span>,</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>               <span class="at">shrd_save_path =</span> <span class="st">'./'</span>,</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>               <span class="at">sample =</span> <span class="st">'CC-HAM-0385'</span>,</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>               <span class="at">plot =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="sc">&gt;</span> shrd_result_borderline</span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a><span class="sc">$</span>hrd_status</span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a>[<span class="dv">1</span>] <span class="cn">FALSE</span></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a><span class="sc">$</span>n_lga</span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a>  Size_LGA Number_LGA</span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a><span class="dv">1</span>        <span class="dv">3</span>         <span class="dv">29</span></span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a><span class="dv">2</span>        <span class="dv">4</span>         <span class="dv">27</span></span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a><span class="dv">3</span>        <span class="dv">5</span>         <span class="dv">27</span></span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a><span class="dv">4</span>        <span class="dv">6</span>         <span class="dv">21</span></span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a><span class="dv">5</span>        <span class="dv">7</span>         <span class="dv">20</span></span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a><span class="dv">6</span>        <span class="dv">8</span>         <span class="dv">19</span></span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a><span class="dv">7</span>        <span class="dv">9</span>         <span class="dv">17</span></span>
<span id="cb7-21"><a href="#cb7-21" tabindex="-1"></a><span class="dv">8</span>       <span class="dv">10</span>         <span class="dv">15</span></span>
<span id="cb7-22"><a href="#cb7-22" tabindex="-1"></a><span class="dv">9</span>       <span class="dv">11</span>         <span class="dv">14</span></span>
<span id="cb7-23"><a href="#cb7-23" tabindex="-1"></a></span>
<span id="cb7-24"><a href="#cb7-24" tabindex="-1"></a><span class="sc">$</span>lga</span>
<span id="cb7-25"><a href="#cb7-25" tabindex="-1"></a>   index chr chr_arm     start       end ratio_median     size level WC</span>
<span id="cb7-26"><a href="#cb7-26" tabindex="-1"></a><span class="dv">3</span>      <span class="dv">3</span>   <span class="dv">1</span>     <span class="fl">1.5</span> <span class="dv">143760001</span> <span class="dv">200670000</span>  <span class="fl">0.051024003</span> <span class="dv">56910000</span>     <span class="dv">3</span>  <span class="dv">0</span></span>
<span id="cb7-27"><a href="#cb7-27" tabindex="-1"></a><span class="dv">4</span>      <span class="dv">4</span>   <span class="dv">1</span>     <span class="fl">1.5</span> <span class="dv">200670001</span> <span class="dv">249210000</span>  <span class="fl">0.387362541</span> <span class="dv">48540000</span>     <span class="dv">4</span>  <span class="dv">1</span></span>
<span id="cb7-28"><a href="#cb7-28" tabindex="-1"></a><span class="dv">8</span>      <span class="dv">8</span>   <span class="dv">3</span>     <span class="fl">3.5</span>  <span class="dv">93540001</span> <span class="dv">130950000</span> <span class="sc">-</span><span class="fl">0.035786412</span> <span class="dv">37410000</span>     <span class="dv">8</span>  <span class="dv">0</span></span>
<span id="cb7-29"><a href="#cb7-29" tabindex="-1"></a><span class="dv">9</span>      <span class="dv">9</span>   <span class="dv">3</span>     <span class="fl">3.5</span> <span class="dv">130950001</span> <span class="dv">147720000</span> <span class="sc">-</span><span class="fl">0.575615328</span> <span class="dv">16770000</span>     <span class="dv">9</span>  <span class="dv">1</span></span>
<span id="cb7-30"><a href="#cb7-30" tabindex="-1"></a><span class="dv">10</span>    <span class="dv">10</span>   <span class="dv">3</span>     <span class="fl">3.5</span> <span class="dv">147720001</span> <span class="dv">197820000</span> <span class="sc">-</span><span class="fl">0.008682243</span> <span class="dv">50100000</span>    <span class="dv">10</span>  <span class="dv">1</span></span>
<span id="cb7-31"><a href="#cb7-31" tabindex="-1"></a><span class="dv">20</span>    <span class="dv">20</span>   <span class="dv">6</span>     <span class="fl">6.5</span>  <span class="dv">61980001</span>  <span class="dv">74970000</span>  <span class="fl">0.353323181</span> <span class="dv">12990000</span>    <span class="dv">20</span>  <span class="dv">0</span></span>
<span id="cb7-32"><a href="#cb7-32" tabindex="-1"></a><span class="dv">21</span>    <span class="dv">21</span>   <span class="dv">6</span>     <span class="fl">6.5</span>  <span class="dv">74970001</span> <span class="dv">170910000</span> <span class="sc">-</span><span class="fl">0.029146346</span> <span class="dv">95940000</span>    <span class="dv">21</span>  <span class="dv">1</span></span>
<span id="cb7-33"><a href="#cb7-33" tabindex="-1"></a><span class="dv">23</span>    <span class="dv">23</span>   <span class="dv">7</span>     <span class="fl">7.5</span>  <span class="dv">62460001</span> <span class="dv">139620000</span>  <span class="fl">0.048236186</span> <span class="dv">77160000</span>    <span class="dv">23</span>  <span class="dv">0</span></span>
<span id="cb7-34"><a href="#cb7-34" tabindex="-1"></a><span class="dv">24</span>    <span class="dv">24</span>   <span class="dv">7</span>     <span class="fl">7.5</span> <span class="dv">139620001</span> <span class="dv">159120000</span>  <span class="fl">0.887525271</span> <span class="dv">19500000</span>    <span class="dv">24</span>  <span class="dv">1</span></span>
<span id="cb7-35"><a href="#cb7-35" tabindex="-1"></a><span class="dv">25</span>    <span class="dv">25</span>   <span class="dv">8</span>     <span class="fl">8.0</span>    <span class="dv">180001</span>  <span class="dv">33360000</span> <span class="sc">-</span><span class="fl">1.514573173</span> <span class="dv">33180000</span>    <span class="dv">25</span>  <span class="dv">0</span></span>
<span id="cb7-36"><a href="#cb7-36" tabindex="-1"></a><span class="dv">26</span>    <span class="dv">26</span>   <span class="dv">8</span>     <span class="fl">8.0</span>  <span class="dv">33360001</span>  <span class="dv">43380000</span>  <span class="fl">0.399444565</span> <span class="dv">10020000</span>    <span class="dv">26</span>  <span class="dv">1</span></span>
<span id="cb7-37"><a href="#cb7-37" tabindex="-1"></a><span class="dv">27</span>    <span class="dv">27</span>   <span class="dv">8</span>     <span class="fl">8.5</span>  <span class="dv">47460001</span>  <span class="dv">67710000</span>  <span class="fl">0.403813062</span> <span class="dv">20250000</span>    <span class="dv">27</span>  <span class="dv">0</span></span>
<span id="cb7-38"><a href="#cb7-38" tabindex="-1"></a><span class="dv">28</span>    <span class="dv">28</span>   <span class="dv">8</span>     <span class="fl">8.5</span>  <span class="dv">67710001</span> <span class="dv">100590000</span>  <span class="fl">0.698218410</span> <span class="dv">32880000</span>    <span class="dv">28</span>  <span class="dv">1</span></span>
<span id="cb7-39"><a href="#cb7-39" tabindex="-1"></a><span class="dv">29</span>    <span class="dv">29</span>   <span class="dv">8</span>     <span class="fl">8.5</span> <span class="dv">100590001</span> <span class="dv">146280000</span>  <span class="fl">0.318461465</span> <span class="dv">45690000</span>    <span class="dv">29</span>  <span class="dv">1</span></span>
<span id="cb7-40"><a href="#cb7-40" tabindex="-1"></a><span class="dv">35</span>    <span class="dv">35</span>  <span class="dv">10</span>    <span class="fl">10.5</span>  <span class="dv">42870001</span>  <span class="dv">61320000</span> <span class="sc">-</span><span class="fl">0.564904848</span> <span class="dv">18450000</span>    <span class="dv">35</span>  <span class="dv">0</span></span>
<span id="cb7-41"><a href="#cb7-41" tabindex="-1"></a><span class="dv">36</span>    <span class="dv">36</span>  <span class="dv">10</span>    <span class="fl">10.5</span>  <span class="dv">61320001</span> <span class="dv">135420000</span>  <span class="fl">0.340277405</span> <span class="dv">74100000</span>    <span class="dv">36</span>  <span class="dv">1</span></span>
<span id="cb7-42"><a href="#cb7-42" tabindex="-1"></a><span class="dv">37</span>    <span class="dv">37</span>  <span class="dv">11</span>    <span class="fl">11.0</span>    <span class="dv">210001</span>  <span class="dv">22860000</span> <span class="sc">-</span><span class="fl">0.575615328</span> <span class="dv">22650000</span>    <span class="dv">37</span>  <span class="dv">0</span></span>
<span id="cb7-43"><a href="#cb7-43" tabindex="-1"></a><span class="dv">38</span>    <span class="dv">38</span>  <span class="dv">11</span>    <span class="fl">11.0</span>  <span class="dv">22860001</span>  <span class="dv">51540000</span> <span class="sc">-</span><span class="fl">0.040971781</span> <span class="dv">28680000</span>    <span class="dv">38</span>  <span class="dv">1</span></span>
<span id="cb7-44"><a href="#cb7-44" tabindex="-1"></a><span class="dv">39</span>    <span class="dv">39</span>  <span class="dv">11</span>    <span class="fl">11.5</span>  <span class="dv">55050001</span>  <span class="dv">81630000</span> <span class="sc">-</span><span class="fl">0.440263476</span> <span class="dv">26580000</span>    <span class="dv">39</span>  <span class="dv">0</span></span>
<span id="cb7-45"><a href="#cb7-45" tabindex="-1"></a><span class="dv">40</span>    <span class="dv">40</span>  <span class="dv">11</span>    <span class="fl">11.5</span>  <span class="dv">81630001</span> <span class="dv">134940000</span> <span class="sc">-</span><span class="fl">0.101598140</span> <span class="dv">53310000</span>    <span class="dv">40</span>  <span class="dv">1</span></span>
<span id="cb7-46"><a href="#cb7-46" tabindex="-1"></a><span class="dv">45</span>    <span class="dv">45</span>  <span class="dv">14</span>    <span class="fl">14.5</span>  <span class="dv">19440001</span>  <span class="dv">73740000</span> <span class="sc">-</span><span class="fl">0.530114400</span> <span class="dv">54300000</span>    <span class="dv">45</span>  <span class="dv">0</span></span>
<span id="cb7-47"><a href="#cb7-47" tabindex="-1"></a><span class="dv">46</span>    <span class="dv">46</span>  <span class="dv">14</span>    <span class="fl">14.5</span>  <span class="dv">73740001</span> <span class="dv">107280000</span> <span class="sc">-</span><span class="fl">0.060397280</span> <span class="dv">33540000</span>    <span class="dv">46</span>  <span class="dv">1</span></span>
<span id="cb7-48"><a href="#cb7-48" tabindex="-1"></a><span class="dv">47</span>    <span class="dv">47</span>  <span class="dv">15</span>    <span class="fl">15.5</span>  <span class="dv">20160001</span>  <span class="dv">91200000</span> <span class="sc">-</span><span class="fl">0.582079992</span> <span class="dv">71040000</span>    <span class="dv">47</span>  <span class="dv">0</span></span>
<span id="cb7-49"><a href="#cb7-49" tabindex="-1"></a><span class="dv">48</span>    <span class="dv">48</span>  <span class="dv">15</span>    <span class="fl">15.5</span>  <span class="dv">91200001</span> <span class="dv">102390000</span>  <span class="fl">0.314406391</span> <span class="dv">11190000</span>    <span class="dv">48</span>  <span class="dv">1</span></span>
<span id="cb7-50"><a href="#cb7-50" tabindex="-1"></a><span class="dv">49</span>    <span class="dv">49</span>  <span class="dv">16</span>    <span class="fl">16.0</span>     <span class="dv">90001</span>  <span class="dv">17610000</span> <span class="sc">-</span><span class="fl">0.384583703</span> <span class="dv">17520000</span>    <span class="dv">49</span>  <span class="dv">0</span></span>
<span id="cb7-51"><a href="#cb7-51" tabindex="-1"></a><span class="dv">50</span>    <span class="dv">50</span>  <span class="dv">16</span>    <span class="fl">16.0</span>  <span class="dv">17610001</span>  <span class="dv">35130000</span>  <span class="fl">0.039840265</span> <span class="dv">17520000</span>    <span class="dv">50</span>  <span class="dv">1</span></span>
<span id="cb7-52"><a href="#cb7-52" tabindex="-1"></a><span class="dv">53</span>    <span class="dv">53</span>  <span class="dv">17</span>    <span class="fl">17.5</span>  <span class="dv">25380001</span>  <span class="dv">48660000</span> <span class="sc">-</span><span class="fl">0.476940288</span> <span class="dv">23280000</span>    <span class="dv">53</span>  <span class="dv">0</span></span>
<span id="cb7-53"><a href="#cb7-53" tabindex="-1"></a><span class="dv">54</span>    <span class="dv">54</span>  <span class="dv">17</span>    <span class="fl">17.5</span>  <span class="dv">48660001</span>  <span class="dv">81090000</span>  <span class="fl">0.037030731</span> <span class="dv">32430000</span>    <span class="dv">54</span>  <span class="dv">1</span></span>
<span id="cb7-54"><a href="#cb7-54" tabindex="-1"></a><span class="dv">70</span>    <span class="dv">70</span>  <span class="dv">23</span>    <span class="fl">23.0</span>    <span class="dv">300001</span>  <span class="dv">16980000</span> <span class="sc">-</span><span class="fl">0.875671865</span> <span class="dv">16680000</span>    <span class="dv">70</span>  <span class="dv">0</span></span>
<span id="cb7-55"><a href="#cb7-55" tabindex="-1"></a><span class="dv">71</span>    <span class="dv">71</span>  <span class="dv">23</span>    <span class="fl">23.0</span>  <span class="dv">16980001</span>  <span class="dv">35760000</span> <span class="sc">-</span><span class="fl">1.457989644</span> <span class="dv">18780000</span>    <span class="dv">71</span>  <span class="dv">1</span></span>
<span id="cb7-56"><a href="#cb7-56" tabindex="-1"></a><span class="dv">72</span>    <span class="dv">72</span>  <span class="dv">23</span>    <span class="fl">23.0</span>  <span class="dv">35760001</span>  <span class="dv">58080000</span> <span class="sc">-</span><span class="fl">0.498178735</span> <span class="dv">22320000</span>    <span class="dv">72</span>  <span class="dv">1</span></span>
<span id="cb7-57"><a href="#cb7-57" tabindex="-1"></a><span class="dv">73</span>    <span class="dv">73</span>  <span class="dv">23</span>    <span class="fl">23.5</span>  <span class="dv">63780001</span> <span class="dv">122790000</span> <span class="sc">-</span><span class="fl">0.529072743</span> <span class="dv">59010000</span>    <span class="dv">73</span>  <span class="dv">0</span></span>
<span id="cb7-58"><a href="#cb7-58" tabindex="-1"></a><span class="dv">74</span>    <span class="dv">74</span>  <span class="dv">23</span>    <span class="fl">23.5</span> <span class="dv">122790001</span> <span class="dv">141840000</span> <span class="sc">-</span><span class="fl">0.064918266</span> <span class="dv">19050000</span>    <span class="dv">74</span>  <span class="dv">1</span></span>
<span id="cb7-59"><a href="#cb7-59" tabindex="-1"></a><span class="dv">75</span>    <span class="dv">75</span>  <span class="dv">23</span>    <span class="fl">23.5</span> <span class="dv">141840001</span> <span class="dv">155220000</span> <span class="sc">-</span><span class="fl">0.655171503</span> <span class="dv">13380000</span>    <span class="dv">75</span>  <span class="dv">1</span></span>
<span id="cb7-60"><a href="#cb7-60" tabindex="-1"></a></span>
<span id="cb7-61"><a href="#cb7-61" tabindex="-1"></a><span class="sc">$</span>plot</span></code></pre></div>
<p>In this sample, we find 15 LGAs, which are visually reflected in the
plot:</p>
<p><img src="images%2Fshallowhrd_vignette_2.png" width="100%"></p>
<div class="section level4 unnumbered">
<h4 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h4>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Callens2023" class="csl-entry">
Callens, Celine, Manuel Rodrigues, Adrien Briaux, Eleonore Frouin,
Alexandre Eeckhoutte, Eric Pujade-Lauraine, Victor Renault, et al. 2023.
<span>“Shallow Whole Genome Sequencing Approach to Detect Homologous
Recombination Deficiency in the PAOLA-1/ENGOT-OV25 Phase-III
Trial.”</span> <em>Oncogene</em> 42 (November): 3556–63. <a href="https://doi.org/10.1038/s41388-023-02839-8" class="external-link">https://doi.org/10.1038/s41388-023-02839-8</a>.
</div>
<div id="ref-Davies2017" class="csl-entry">
Davies, Helen, Dominik Glodzik, Sandro Morganella, Lucy R Yates, Johan
Staaf, Xueqing Zou, Manasa Ramakrishna, et al. 2017. <span>“HRDetect Is
a Predictor of BRCA1 and BRCA2 Deficiency Based on Mutational
Signatures.”</span> <em>Nature Medicine</em> 23 (April): 517–25. <a href="https://doi.org/10.1038/nm.4292" class="external-link">https://doi.org/10.1038/nm.4292</a>.
</div>
<div id="ref-Eeckhoutte2020" class="csl-entry">
Eeckhoutte, Alexandre, Alexandre Houy, Elodie Manié, Manon Reverdy, Ivan
Bièche, Elisabetta Marangoni, Oumou Goundiam, et al. 2020.
<span>“ShallowHRD: Detection of Homologous Recombination Deficiency from
Shallow Whole Genome Sequencing.”</span> <em>Bioinformatics</em> 36
(June): 3888–89. <a href="https://doi.org/10.1093/bioinformatics/btaa261" class="external-link">https://doi.org/10.1093/bioinformatics/btaa261</a>.
</div>
<div id="ref-Guo2023" class="csl-entry">
Guo, Yuanli, Xinxin He, Yanming Tan, Junfeng Liu, Huadong Chen, Yi
Huang, Chao Zhang, Ying Tao, and Shan Chen. 2023. <span>“ShallowHRD
Status Acts as an Effective Prognostic Predictor in Ovarian Cancer
Patients Treated by Poly (ADP-Ribose) Polymerase Inhibitors
(PARPis).”</span> <em>Journal of Cancer Research and Clinical
Oncology</em> 149 (November): 15839–44. <a href="https://doi.org/10.1007/s00432-023-05341-6" class="external-link">https://doi.org/10.1007/s00432-023-05341-6</a>.
</div>
<div id="ref-Jamieson2024" class="csl-entry">
Jamieson, Amy, Juliana Sobral de Barros, Dawn R. Cochrane, J. Maxwell
Douglas, Sameer Shankar, Branden J. Lynch, Samuel Leung, et al. 2024.
<span>“Targeted and Shallow Whole-Genome Sequencing Identifies
Therapeutic Opportunities in P53abn Endometrial Cancers.”</span>
<em>Clinical Cancer Research</em>, April, OF1–14. <a href="https://doi.org/10.1158/1078-0432.CCR-23-3689" class="external-link">https://doi.org/10.1158/1078-0432.CCR-23-3689</a>.
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Maxwell Douglas, Branden Lynch, Carlos Vasquez, Jacky Yiu, Matthew Eldridge, Geoff Macintyre, Ding Ma, Nirupama Tamvada, Sameer Shankar.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
